# Import section 

In [13]:
import pandas as pd
import numpy as np
from pprint import pprint
import requests
from bs4 import BeautifulSoup
import requests, re, json
import random
from tqdm import tqdm
from googlemaps import Client as GoogleMaps
import googlemaps
import gmaps
from keplergl import KeplerGl
import geopandas as gpd
import matplotlib.pyplot as plt
import time
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager
pd.set_option('mode.chained_assignment', None)

# Intro

Within this project i will try to evaluate different factors to predict the rent of apartments and houses in London. The whole project is separate into four sections: 
1. Scaring the real estate Data from Zoopla
2. Scraping additional information via api and webscraping
3. EDA of the collected data
4. Modeling 

Within this notebook i will focus on scraping and cleaning the data from Zoopla 

# Functions and Information for Scraping

## User agent switching:

Within the first attempts of the scraping i realized that Zoopla will block my requests after some time. The website is identifying that it is scraped by a program. to avoid this i will create a functions which is changing the user agent details of my requests. Like this it will look like that a real person is requesting the data. 

In [14]:
# creating a dictionry with the fake headers
headers_dict = {0 : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36",
               1 : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36",
               2 : "Safari/13605.3.8 CFNetwork/902.1 Darwin/17.7.0 (x86_64)",
               3 : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.80 Safari/537.36",
               4 : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:59.0) Gecko/20100101 Firefox/59.0"
               }

In [15]:
# the function will randomly choose one of the headers and return it
def generate_random_header():
    random_agent = random.choice(headers_dict)
    header = headers = {'User-Agent': random_agent}
    return header

## Creation of URLs


First I will have a look at the first three Zoopla pages for the search in London to identify the difference.

https://www.zoopla.co.uk/to-rent/property/london/?price_frequency=per_month&q=London&results_sort=newest_listings&search_source=to-rent
https://www.zoopla.co.uk/to-rent/property/london/?price_frequency=per_month&q=London&results_sort=newest_listings&search_source=to-rent&pn=2
https://www.zoopla.co.uk/to-rent/property/london/?price_frequency=per_month&q=London&results_sort=newest_listings&search_source=to-rent&pn=3

The only difference is the &pn=2 at the end. this indicates which side of the search result it is. On the website we can see that there are 400 pages with 25 results per page. This would lead to a total of 10.000 results. Nevertheless on the website it is stating that there are 10.000+ results. I assume there are much more results available but the website is only capable to show 10,000 results. 

After investigating different price categories i realized that already for a price range between 1000 and 2000 per month I get more than 10.000 results. So we would need to create more links for the different price sections. I will look at two different results to identify how the min an max price need to be included. 

https://www.zoopla.co.uk/to-rent/property/london/?page_size=25&price_frequency=per_month&price_max=2000&price_min=1000&view_type=list&q=London&radius=0&results_sort=newest_listings&search_source=refine
https://www.zoopla.co.uk/to-rent/property/london/?page_size=25&price_frequency=per_month&price_max=2250&price_min=2000&view_type=list&q=London&radius=0&results_sort=newest_listings&search_source=refine
https://www.zoopla.co.uk/to-rent/property/london/?page_size=25&price_frequency=per_month&price_max=25000&price_min=4000&view_type=list&q=London&radius=0&results_sort=newest_listings&search_source=refine

Afterwards i will create a dictionary with the different price ranges. To define the ranges i am checking on the actually website and always try to stay below 8.000 results. Like this there is some buffer if more objects are published to the market within one of the ranges. The goal is to always be able to scrape all of the data.  

In [16]:
zoopla_pricecategories =    {0:["","&price_max=1500"],
                             1:["&price_min1500", "&price_max=2000"],
                             2:["&price_min2000", "&price_max=2750"],
                             3:["&price_min2750", "&price_max=4000"],
                             4:["&price_min4000", ""],
                             }

The problem is we don't know now how many pages are existing for every section. But we know the maximum is 400. So we will create for each sections links for 400 pages. Afterwards within the scraping function we will include a try-and-except-statement to skip a side if it is not available.  

In [17]:
zoopla_links =[]

for key,value in zoopla_pricecategories.items():
    for n in range(1,151):
        max_price = value[1]
        min_price = value[0] 
        if n == 1:
            index = ""
        else:
            index = f'&pn={n}'
        url=f'https://www.zoopla.co.uk/to-rent/property/london/?page_size=25&price_frequency=per_month{max_price}{min_price}&view_type=list&q=London&radius=0&results_sort=newest_listings&search_source=refine{index}'
        zoopla_links.append(url)

**Attention**

At a later stage i will run the scraper again to collect additional real estate data. For the second run i will reduce the amount of pages to 150. The Zoopla website is sorting the results always by newest first. I am planning to run the scraper every month. like this i should always collect the new published objects and i can increase my dataset. 

In [18]:
len(zoopla_links)

750

Within the first run the loop create 2000 links which we will use for the scrapping. For the updates the amount of pages for the scraping was reduced as i only want to include the newest results. For this reason it only shows 750 results.

## Function to scrape real estate data

First I will create an empty dataFrame to store the scraped data. Within the name i am actually including the date. like this I can keep the results separated and combine the dataFrames afterwards. 

In [19]:
df_0711_zoopla = pd.DataFrame(columns =["name", "location", "price", "bedrooms", "bathrooms", "receptions", "published", "seize", "public_transport", "describtion", "links", "latitude", "longitude"])

In case my scraper is crashing in between I am setting up am empty list which is storing the links that had been scraped. In case of a crash i can run the scraper again from the last link and don't have to run everything again

In [20]:
already_done =[]

**Function for the scraping**

In a short summary the function includes several steps to scrape the data from Zoopla. First the available data from the search result will be exported. Afterwards the scraper will open the Zoopla page for each object to scrape additional information. Mainly i will use Beautifulsoup to extract the data and regex commands to extract the different information.  After one link (25 Objects) has been scraped that the data will automatically be saved locally on my machine. More details are explained within the comments in the function.

In [21]:
def scraping_zoopla():
    #seting the DataFrame as global so that it can be used within the function
    global df_0711_zoopla

    for i in tqdm(zoopla_links): 
        
        #1. choosing url:
        
        url = i
        already_done.append(url) # in case of crash of scraper to keep track
        #print("1 ----------------------------------------------------------------")
        
        #2. generating a random header
        
        random_header = generate_random_header()
        #print(random_header) # print header to identify if one is not working
        
        #print("2 ----------------------------------------------------------------")
        
        #3. request the page and transform the content with beautifulsoup
        
        page = requests.get(url, headers=random_header)
        content = BeautifulSoup(page.content, "lxml")

        #print("3 ----------------------------------------------------------------")

        #4. setting up the empty lists for the informaiton     

        price=[]
        location=[]
        name=[]
        published = []
        public_transport = []
        all_apartment_links = []
        columns =["name", "location", "price", "bedrooms", "bathrooms", "receptions",
                  "published", "seize", "public_transport", "describtion", "links", 
                  "latitude", "longitude"]

        #print("4 ----------------------------------------------------------------")
        
        #5. looping through the page to extract the content
        for i in content.findAll('div',{'data-testid':'search-result'}):
    
            #scraping of price
            for j in i.findAll("div", {"data-testid":"listing-price"}): 
                price.append(j.text.strip())

            # scraping of name/title    
            for j in i.findAll("h2", {"data-testid":"listing-title"}): 
                name.append(j.text.strip())
        
            # scraping of location   
            for j in i.findAll("p", {"data-testid":"listing-description"}): 
                location.append(j.text.strip()) 
    
            # scraping of public transportation
            for j in i.findAll("div", {"data-testid":"listing-transport"}): 
                public_transport.append(j.text.strip()) 
    
            # scraping of publishing date
            for j in i.findAll("span", {"data-testid":"date-published"}): 
                published.append(j.text.strip())
        
            # scraping of links also for additional information
            for j in i.findAll("a", {"data-testid": "listing-details-link"}):
                apartment_info = j.get('href')
                link = "https://www.zoopla.co.uk" + apartment_info
                all_apartment_links.append(link)
        
        #print("5 ----------------------------------------------------------------")  
        
        #6. scraping additional informaiton from the single links for more information see teh function additional_info
        all_apartment_links = list(dict.fromkeys(all_apartment_links))
        additional_info = scraping_additional_info(all_apartment_links)   
    
        bedrooms = additional_info[0]
        bathrooms = additional_info[1]
        receptions = additional_info[2]
        seize = additional_info[3]
        describtion = additional_info[4]
        longitude = additional_info[5]
        latitude = additional_info[6]
                    
        #print("6 ----------------------------------------------------------------") 
        
        #7. adding information to a DataFrame 
        zipped = list(zip(name, location, price, bedrooms, bathrooms, receptions, published,
                          seize, public_transport, describtion, all_apartment_links, latitude, longitude))
        df_temp=pd.DataFrame(zipped, columns=columns)
        
        df_0711_zoopla = pd.concat([df_0711_zoopla, df_temp], axis=0, ignore_index = True)
        
        print(df_0711_zoopla.shape) # control print of the shape to see if zoopla is blocking my requests
        
        #print("7 ----------------------------------------------------------------") 
        
        #8. at the end of each iteration save the panda
            
        df_0711_zoopla.to_csv(f'df_0711_zoopla.csv')
        
        time.sleep(4) #the code will stop for 4 seconds like this i can avoid captures
        #print("8 ----------------------------------------------------------------")     

I realized that not all information are directly available on the search result page of Zoopla. Bedrooms, bathrooms, description and other information are stored on the single sides of each object. For this reason is created a second function to scrape the additional data from each object page. Also here I used BeautifoulSoup and regex to extract the information. More information are included as comments in the function. 

**Function to scrape additional information for each object**

In [10]:
def scraping_additional_info(links_from_properties):
    #creation of empty lists to store the data
    bedrooms = []
    bathrooms = []
    receptions = []
    seize = []
    describtion = []
    longitude = []
    latitude = []
    
    #iteartion through each single object page that ahs been collected on the zoopla result page
    for link in links_from_properties:
        
        # Generating a random header
        header = generate_random_header()
        details = requests.get(link,header)
        #Extracting and transforming the html 
        details_appartment = BeautifulSoup(details.content, "lxml")
        text = details_appartment.text.strip()
        
        # Extracting the amount of bedrooms
        try:
            bedrooms.append(re.search(r"[0-9].(beds|bed)", text)[0])
        except:
            bedrooms.append(np.NAN)
        
        # Extracting the amount of bathrooms
        try:
            bathrooms.append(re.search(r"[0-9].(baths|bath)", text)[0])
        except:
            bathrooms.append(np.NAN) 
        
        # Extracting the amount of receptions
        try:
            receptions.append(re.search(r"[0-9].(reception|receptions)", text)[0])
        except:
            receptions.append(np.NAN)
            
        # Extracting the seize of the object    
        try:
            seize.append(re.search(r"[0-9|,]*?.sq. ft", text)[0])
        except:
            seize.append(np.NAN)   
            
        # extracting the describtion of the object    
        try:
            try:
                describtion.append(re.search(r"Features and description.+?(?=Floor plans and tours)", text)[0])
            except:
                describtion.append(re.search(r"Features and description.+?(?=More information)", text)[0])
        except:
            describtion.append(np.NAN)
            
        # extracting the longitude and latitude to locate the object    
        try:
            longitude.append(float(re.search(r"(?<=.longitude.: )[-0-9]*.[0-9]*", str(details.content))[0]))
            latitude.append(float(re.search(r"(?<=.latitude.: )[-0-9]*.[0-9]*", str(details.content))[0]))
        except:
            longitude.append(np.NAN)
            latitude.append(np.NAN)
         
        # i included a sleep timer of 4 seconds after every iteration to avoid captures
        time.sleep(4)    
        
        #at the end the function will return the scraped data
    return bedrooms, bathrooms, receptions, seize , describtion, longitude, latitude    

# Scraping Zoopla

In [22]:
scraping_zoopla()

  0%|                                                   | 0/750 [00:00<?, ?it/s]

(25, 13)


  0%|                                       | 1/750 [02:10<27:06:17, 130.28s/it]

(50, 13)


  0%|                                       | 2/750 [04:21<27:08:59, 130.67s/it]

(75, 13)


  0%|▏                                      | 3/750 [06:29<26:55:08, 129.73s/it]

(100, 13)


  1%|▏                                      | 4/750 [08:40<26:56:08, 129.98s/it]

(125, 13)


  1%|▎                                      | 5/750 [10:49<26:48:42, 129.56s/it]

(150, 13)


  1%|▎                                      | 6/750 [13:04<27:13:14, 131.71s/it]

(175, 13)


  1%|▎                                      | 7/750 [15:14<27:01:43, 130.96s/it]

(200, 13)


  1%|▍                                      | 8/750 [17:20<26:41:29, 129.50s/it]

(225, 13)


  1%|▍                                      | 9/750 [19:28<26:34:35, 129.12s/it]

(250, 13)


  1%|▌                                     | 10/750 [21:36<26:24:35, 128.48s/it]

(275, 13)


  1%|▌                                     | 11/750 [23:42<26:16:25, 127.99s/it]

(300, 13)


  2%|▌                                     | 12/750 [25:50<26:13:47, 127.95s/it]

(325, 13)


  2%|▋                                     | 13/750 [27:57<26:07:37, 127.62s/it]

(350, 13)


  2%|▋                                     | 14/750 [30:03<26:00:24, 127.21s/it]

(375, 13)


  2%|▊                                     | 15/750 [32:12<26:05:21, 127.78s/it]

(400, 13)


  2%|▊                                     | 16/750 [34:20<26:02:13, 127.70s/it]

(425, 13)


  2%|▊                                     | 17/750 [36:30<26:07:30, 128.31s/it]

(450, 13)


  2%|▉                                     | 18/750 [38:40<26:14:03, 129.02s/it]

(475, 13)


  3%|▉                                     | 19/750 [40:49<26:10:17, 128.89s/it]

(500, 13)


  3%|█                                     | 20/750 [42:59<26:12:24, 129.24s/it]

(525, 13)


  3%|█                                     | 21/750 [45:06<26:01:48, 128.54s/it]

(550, 13)


  3%|█                                     | 22/750 [47:12<25:51:39, 127.88s/it]

(575, 13)


  3%|█▏                                    | 23/750 [49:20<25:47:16, 127.70s/it]

(600, 13)


  3%|█▏                                    | 24/750 [51:25<25:38:26, 127.14s/it]

(625, 13)


  3%|█▎                                    | 25/750 [53:32<25:32:38, 126.84s/it]

(650, 13)


  3%|█▎                                    | 26/750 [55:40<25:37:42, 127.43s/it]

(675, 13)


  4%|█▎                                    | 27/750 [57:47<25:32:28, 127.18s/it]

(700, 13)


  4%|█▍                                    | 28/750 [59:53<25:25:59, 126.81s/it]

(725, 13)


  4%|█▍                                  | 29/750 [1:01:58<25:16:29, 126.20s/it]

(750, 13)


  4%|█▍                                  | 30/750 [1:04:03<25:09:55, 125.83s/it]

(775, 13)


  4%|█▍                                  | 31/750 [1:06:10<25:14:13, 126.36s/it]

(800, 13)


  4%|█▌                                  | 32/750 [1:08:17<25:14:02, 126.52s/it]

(825, 13)


  4%|█▌                                  | 33/750 [1:10:24<25:12:33, 126.57s/it]

(850, 13)


  5%|█▋                                  | 34/750 [1:12:31<25:13:32, 126.83s/it]

(875, 13)


  5%|█▋                                  | 35/750 [1:14:39<25:13:16, 126.99s/it]

(900, 13)


  5%|█▋                                  | 36/750 [1:16:46<25:13:06, 127.15s/it]

(925, 13)


  5%|█▊                                  | 37/750 [1:18:52<25:06:20, 126.76s/it]

(950, 13)


  5%|█▊                                  | 38/750 [1:20:59<25:06:00, 126.91s/it]

(975, 13)


  5%|█▊                                  | 39/750 [1:23:09<25:13:37, 127.73s/it]

(1000, 13)


  5%|█▉                                  | 40/750 [1:25:19<25:20:16, 128.47s/it]

(1025, 13)


  5%|█▉                                  | 41/750 [1:27:25<25:10:07, 127.80s/it]

(1050, 13)


  6%|██                                  | 42/750 [1:29:32<25:04:32, 127.50s/it]

(1075, 13)


  6%|██                                  | 43/750 [1:31:42<25:09:22, 128.09s/it]

(1100, 13)


  6%|██                                  | 44/750 [1:33:49<25:06:24, 128.02s/it]

(1125, 13)


  6%|██▏                                 | 45/750 [1:35:56<25:00:14, 127.68s/it]

(1150, 13)


  6%|██▏                                 | 46/750 [1:38:06<25:05:56, 128.35s/it]

(1175, 13)


  6%|██▎                                 | 47/750 [1:40:14<25:01:48, 128.18s/it]

(1200, 13)


  6%|██▎                                 | 48/750 [1:42:20<24:53:14, 127.63s/it]

(1225, 13)


  7%|██▎                                 | 49/750 [1:44:28<24:52:29, 127.75s/it]

(1250, 13)


  7%|██▍                                 | 50/750 [1:46:35<24:45:57, 127.37s/it]

(1275, 13)


  7%|██▍                                 | 51/750 [1:48:41<24:38:36, 126.92s/it]

(1300, 13)


  7%|██▍                                 | 52/750 [1:50:49<24:40:55, 127.30s/it]

(1325, 13)


  7%|██▌                                 | 53/750 [1:52:56<24:36:48, 127.13s/it]

(1350, 13)


  7%|██▌                                 | 54/750 [1:55:05<24:42:50, 127.83s/it]

(1375, 13)


  7%|██▋                                 | 55/750 [1:57:12<24:38:56, 127.68s/it]

(1400, 13)


  7%|██▋                                 | 56/750 [1:59:19<24:33:39, 127.41s/it]

(1425, 13)


  8%|██▋                                 | 57/750 [2:01:27<24:32:43, 127.51s/it]

(1450, 13)


  8%|██▊                                 | 58/750 [2:03:34<24:27:29, 127.24s/it]

(1475, 13)


  8%|██▊                                 | 59/750 [2:05:40<24:22:08, 126.96s/it]

(1500, 13)


  8%|██▉                                 | 60/750 [2:07:48<24:23:06, 127.23s/it]

(1525, 13)


  8%|██▉                                 | 61/750 [2:09:54<24:17:41, 126.94s/it]

(1550, 13)


  8%|██▉                                 | 62/750 [2:12:02<24:20:22, 127.36s/it]

(1575, 13)


  8%|███                                 | 63/750 [2:14:12<24:24:19, 127.89s/it]

(1600, 13)


  9%|███                                 | 64/750 [2:16:20<24:25:22, 128.17s/it]

(1625, 13)


  9%|███                                 | 65/750 [2:18:29<24:24:11, 128.25s/it]

(1650, 13)


  9%|███▏                                | 66/750 [2:20:37<24:20:54, 128.15s/it]

(1675, 13)


  9%|███▏                                | 67/750 [2:22:45<24:19:54, 128.25s/it]

(1700, 13)


  9%|███▎                                | 68/750 [2:24:54<24:20:46, 128.51s/it]

(1725, 13)


  9%|███▎                                | 69/750 [2:27:04<24:22:30, 128.86s/it]

(1750, 13)


  9%|███▎                                | 70/750 [2:29:23<24:53:46, 131.80s/it]

(1775, 13)


  9%|███▍                                | 71/750 [2:31:30<24:37:28, 130.56s/it]

(1800, 13)


 10%|███▍                                | 72/750 [2:33:40<24:32:34, 130.32s/it]

(1825, 13)


 10%|███▌                                | 73/750 [2:35:49<24:24:53, 129.83s/it]

(1850, 13)


 10%|███▌                                | 74/750 [2:37:59<24:23:14, 129.87s/it]

(1875, 13)


 10%|███▌                                | 75/750 [2:40:09<24:20:55, 129.86s/it]

(1900, 13)


 10%|███▋                                | 76/750 [2:42:19<24:22:26, 130.19s/it]

(1925, 13)


 10%|███▋                                | 77/750 [2:44:29<24:16:44, 129.87s/it]

(1950, 13)


 10%|███▋                                | 78/750 [2:46:38<24:12:02, 129.65s/it]

(1975, 13)


 11%|███▊                                | 79/750 [2:48:55<24:34:56, 131.89s/it]

(2000, 13)


 11%|███▊                                | 80/750 [2:51:11<24:45:44, 133.05s/it]

(2025, 13)


 11%|███▉                                | 81/750 [2:53:27<24:54:12, 134.01s/it]

(2050, 13)


 11%|███▉                                | 82/750 [2:55:42<24:55:21, 134.31s/it]

(2075, 13)


 11%|███▉                                | 83/750 [2:57:57<24:55:11, 134.50s/it]

(2100, 13)


 11%|████                                | 84/750 [3:00:10<24:46:54, 133.96s/it]

(2125, 13)


 11%|████                                | 85/750 [3:02:24<24:45:46, 134.05s/it]

(2150, 13)


 11%|████▏                               | 86/750 [3:04:39<24:48:53, 134.54s/it]

(2175, 13)


 12%|████▏                               | 87/750 [3:06:55<24:50:13, 134.86s/it]

(2200, 13)


 12%|████▏                               | 88/750 [3:09:10<24:46:45, 134.75s/it]

(2225, 13)


 12%|████▎                               | 89/750 [3:11:25<24:46:35, 134.94s/it]

(2250, 13)


 12%|████▎                               | 90/750 [3:13:32<24:19:17, 132.66s/it]

(2275, 13)


 12%|████▎                               | 91/750 [3:15:43<24:10:59, 132.11s/it]

(2300, 13)


 12%|████▍                               | 92/750 [3:17:50<23:52:00, 130.58s/it]

(2325, 13)


 12%|████▍                               | 93/750 [3:20:01<23:49:28, 130.55s/it]

(2350, 13)


 13%|████▌                               | 94/750 [3:22:10<23:42:25, 130.10s/it]

(2375, 13)


 13%|████▌                               | 95/750 [3:24:18<23:35:00, 129.62s/it]

(2400, 13)


 13%|████▌                               | 96/750 [3:26:28<23:32:35, 129.60s/it]

(2425, 13)


 13%|████▋                               | 97/750 [3:28:36<23:25:33, 129.15s/it]

(2450, 13)


 13%|████▋                               | 98/750 [3:30:43<23:18:10, 128.67s/it]

(2475, 13)


 13%|████▊                               | 99/750 [3:32:53<23:19:33, 128.99s/it]

(2500, 13)


 13%|████▋                              | 100/750 [3:35:04<23:24:01, 129.60s/it]

(2525, 13)


 13%|████▋                              | 101/750 [3:37:15<23:24:35, 129.85s/it]

(2550, 13)


 14%|████▊                              | 102/750 [3:39:28<23:33:03, 130.84s/it]

(2575, 13)


 14%|████▊                              | 103/750 [3:41:36<23:23:58, 130.20s/it]

(2600, 13)


 14%|████▊                              | 104/750 [3:43:43<23:09:37, 129.07s/it]

(2625, 13)


 14%|████▉                              | 105/750 [3:45:50<23:00:44, 128.44s/it]

(2650, 13)


 14%|████▉                              | 106/750 [3:48:05<23:19:25, 130.38s/it]

(2675, 13)


 14%|████▉                              | 107/750 [3:50:24<23:45:34, 133.02s/it]

(2700, 13)


 14%|█████                              | 108/750 [3:52:35<23:36:55, 132.42s/it]

(2725, 13)


 15%|█████                              | 109/750 [3:54:48<23:35:25, 132.49s/it]

(2750, 13)


 15%|█████▏                             | 110/750 [3:56:59<23:30:45, 132.26s/it]

(2775, 13)


 15%|█████▏                             | 111/750 [3:59:13<23:33:32, 132.73s/it]

(2800, 13)


 15%|█████▏                             | 112/750 [4:01:26<23:32:45, 132.86s/it]

(2825, 13)


 15%|█████▎                             | 113/750 [4:03:38<23:27:09, 132.54s/it]

(2850, 13)


 15%|█████▎                             | 114/750 [4:05:52<23:28:07, 132.84s/it]

(2875, 13)


 15%|█████▎                             | 115/750 [4:08:07<23:34:50, 133.69s/it]

(2900, 13)


 15%|█████▍                             | 116/750 [4:10:20<23:30:36, 133.50s/it]

(2925, 13)


 16%|█████▍                             | 117/750 [4:12:29<23:13:21, 132.07s/it]

(2950, 13)


 16%|█████▌                             | 118/750 [4:14:36<22:54:18, 130.47s/it]

(2975, 13)


 16%|█████▌                             | 119/750 [4:16:43<22:42:34, 129.56s/it]

(3000, 13)


 16%|█████▌                             | 120/750 [4:18:52<22:37:19, 129.27s/it]

(3025, 13)


 16%|█████▋                             | 121/750 [4:21:00<22:32:16, 128.99s/it]

(3050, 13)


 16%|█████▋                             | 122/750 [4:23:11<22:34:15, 129.39s/it]

(3075, 13)


 16%|█████▋                             | 123/750 [4:25:25<22:47:17, 130.84s/it]

(3100, 13)


 17%|█████▊                             | 124/750 [4:27:40<22:59:34, 132.23s/it]

(3125, 13)


 17%|█████▊                             | 125/750 [4:29:53<23:00:10, 132.50s/it]

(3150, 13)


 17%|█████▉                             | 126/750 [4:32:09<23:08:26, 133.50s/it]

(3175, 13)


 17%|█████▉                             | 127/750 [4:34:21<23:00:09, 132.92s/it]

(3200, 13)


 17%|█████▉                             | 128/750 [4:36:37<23:07:45, 133.87s/it]

(3225, 13)


 17%|██████                             | 129/750 [4:38:55<23:18:44, 135.14s/it]

(3250, 13)


 17%|██████                             | 130/750 [4:41:09<23:13:45, 134.88s/it]

(3275, 13)


 17%|██████                             | 131/750 [4:43:22<23:04:01, 134.15s/it]

(3300, 13)


 18%|██████▏                            | 132/750 [4:45:35<22:59:48, 133.96s/it]

(3325, 13)


 18%|██████▏                            | 133/750 [4:47:53<23:10:47, 135.25s/it]

(3350, 13)


 18%|██████▎                            | 134/750 [4:50:08<23:06:13, 135.02s/it]

(3375, 13)


 18%|██████▎                            | 135/750 [4:52:24<23:06:34, 135.27s/it]

(3400, 13)


 18%|██████▎                            | 136/750 [4:54:39<23:04:09, 135.26s/it]

(3425, 13)


 18%|██████▍                            | 137/750 [4:56:54<23:01:03, 135.18s/it]

(3450, 13)


 18%|██████▍                            | 138/750 [4:59:09<22:58:15, 135.12s/it]

(3475, 13)


 19%|██████▍                            | 139/750 [5:01:24<22:55:40, 135.09s/it]

(3500, 13)


 19%|██████▌                            | 140/750 [5:03:38<22:49:30, 134.71s/it]

(3525, 13)


 19%|██████▌                            | 141/750 [5:05:57<23:02:15, 136.18s/it]

(3550, 13)


 19%|██████▋                            | 142/750 [5:08:12<22:55:17, 135.72s/it]

(3575, 13)


 19%|██████▋                            | 143/750 [5:10:29<22:55:58, 136.01s/it]

(3600, 13)


 19%|██████▋                            | 144/750 [5:12:37<22:28:54, 133.56s/it]

(3625, 13)


 19%|██████▊                            | 145/750 [5:14:45<22:11:56, 132.09s/it]

(3650, 13)


 19%|██████▊                            | 146/750 [5:16:53<21:55:45, 130.70s/it]

(3675, 13)


 20%|██████▊                            | 147/750 [5:19:02<21:47:57, 130.14s/it]

(3700, 13)


 20%|██████▉                            | 148/750 [5:21:11<21:44:43, 130.04s/it]

(3725, 13)


 20%|██████▉                            | 149/750 [5:23:28<22:01:10, 131.90s/it]

(3750, 13)


 20%|███████                            | 150/750 [5:25:43<22:07:59, 132.80s/it]

(3775, 13)


 20%|███████                            | 151/750 [5:27:57<22:10:40, 133.29s/it]

(3800, 13)


 20%|███████                            | 152/750 [5:30:10<22:07:00, 133.15s/it]

(3825, 13)


 20%|███████▏                           | 153/750 [5:32:23<22:03:52, 133.05s/it]

(3850, 13)


 21%|███████▏                           | 154/750 [5:34:36<22:04:02, 133.29s/it]

(3875, 13)


 21%|███████▏                           | 155/750 [5:36:49<22:00:30, 133.16s/it]

(3900, 13)


 21%|███████▎                           | 156/750 [5:39:03<21:59:03, 133.24s/it]

(3925, 13)


 21%|███████▎                           | 157/750 [5:41:18<22:01:30, 133.71s/it]

(3950, 13)


 21%|███████▎                           | 158/750 [5:43:30<21:55:44, 133.35s/it]

(3975, 13)


 21%|███████▍                           | 159/750 [5:45:43<21:52:52, 133.29s/it]

(4000, 13)


 21%|███████▍                           | 160/750 [5:47:56<21:49:10, 133.14s/it]

(4025, 13)


 21%|███████▌                           | 161/750 [5:50:08<21:42:37, 132.69s/it]

(4050, 13)


 22%|███████▌                           | 162/750 [5:52:22<21:45:51, 133.25s/it]

(4075, 13)


 22%|███████▌                           | 163/750 [5:54:34<21:39:14, 132.80s/it]

(4100, 13)


 22%|███████▋                           | 164/750 [5:56:47<21:38:51, 132.99s/it]

(4125, 13)


 22%|███████▋                           | 165/750 [5:59:00<21:36:38, 132.99s/it]

(4150, 13)


 22%|███████▋                           | 166/750 [6:01:12<21:30:35, 132.60s/it]

(4175, 13)


 22%|███████▊                           | 167/750 [6:03:26<21:31:07, 132.88s/it]

(4200, 13)


 22%|███████▊                           | 168/750 [6:05:39<21:30:52, 133.08s/it]

(4225, 13)


 23%|███████▉                           | 169/750 [6:07:51<21:24:44, 132.68s/it]

(4250, 13)


 23%|███████▉                           | 170/750 [6:10:02<21:19:20, 132.35s/it]

(4275, 13)


 23%|███████▉                           | 171/750 [6:12:21<21:33:52, 134.08s/it]

(4300, 13)


 23%|████████                           | 172/750 [6:14:36<21:35:18, 134.46s/it]

(4325, 13)


 23%|████████                           | 173/750 [6:16:47<21:23:22, 133.45s/it]

(4350, 13)


 23%|████████                           | 174/750 [6:18:58<21:13:40, 132.67s/it]

(4375, 13)


 23%|████████▏                          | 175/750 [6:21:09<21:07:16, 132.24s/it]

(4400, 13)


 23%|████████▏                          | 176/750 [6:23:22<21:06:12, 132.36s/it]

(4425, 13)


 24%|████████▎                          | 177/750 [6:25:33<21:01:10, 132.06s/it]

(4450, 13)


 24%|████████▎                          | 178/750 [6:27:44<20:56:06, 131.76s/it]

(4475, 13)


 24%|████████▎                          | 179/750 [6:29:59<21:01:51, 132.59s/it]

(4500, 13)


 24%|████████▍                          | 180/750 [6:32:10<20:55:28, 132.15s/it]

(4525, 13)


 24%|████████▍                          | 181/750 [6:34:21<20:51:56, 132.02s/it]

(4550, 13)


 24%|████████▍                          | 182/750 [6:36:34<20:50:51, 132.13s/it]

(4575, 13)


 24%|████████▌                          | 183/750 [6:38:44<20:43:17, 131.57s/it]

(4600, 13)


 25%|████████▌                          | 184/750 [6:40:56<20:41:54, 131.65s/it]

(4625, 13)


 25%|████████▋                          | 185/750 [6:43:09<20:43:34, 132.06s/it]

(4650, 13)


 25%|████████▋                          | 186/750 [6:45:23<20:46:58, 132.66s/it]

(4675, 13)


 25%|████████▋                          | 187/750 [6:47:36<20:46:51, 132.88s/it]

(4700, 13)


 25%|████████▊                          | 188/750 [6:49:46<20:34:11, 131.76s/it]

(4725, 13)


 25%|████████▊                          | 189/750 [6:51:58<20:32:33, 131.82s/it]

(4750, 13)


 25%|████████▊                          | 190/750 [6:54:11<20:35:03, 132.33s/it]

(4775, 13)


 25%|████████▉                          | 191/750 [6:56:22<20:30:12, 132.04s/it]

(4800, 13)


 26%|████████▉                          | 192/750 [6:58:33<20:24:47, 131.70s/it]

(4825, 13)


 26%|█████████                          | 193/750 [7:00:47<20:28:25, 132.33s/it]

(4850, 13)


 26%|█████████                          | 194/750 [7:03:00<20:27:20, 132.45s/it]

(4875, 13)


 26%|█████████                          | 195/750 [7:05:14<20:29:28, 132.92s/it]

(4900, 13)


 26%|█████████▏                         | 196/750 [7:07:27<20:26:34, 132.84s/it]

(4925, 13)


 26%|█████████▏                         | 197/750 [7:09:39<20:24:03, 132.81s/it]

(4950, 13)


 26%|█████████▏                         | 198/750 [7:11:55<20:29:53, 133.68s/it]

(4975, 13)


 27%|█████████▎                         | 199/750 [7:14:10<20:29:51, 133.92s/it]

(5000, 13)


 27%|█████████▎                         | 200/750 [7:16:33<20:53:17, 136.72s/it]

(5025, 13)


 27%|█████████▍                         | 201/750 [7:18:53<21:00:54, 137.80s/it]

(5050, 13)


 27%|█████████▍                         | 202/750 [7:21:06<20:45:23, 136.36s/it]

(5075, 13)


 27%|█████████▍                         | 203/750 [7:23:19<20:32:39, 135.21s/it]

(5100, 13)


 27%|█████████▌                         | 204/750 [7:25:31<20:23:19, 134.43s/it]

(5125, 13)


 27%|█████████▌                         | 205/750 [7:27:45<20:19:38, 134.27s/it]

(5150, 13)


 27%|█████████▌                         | 206/750 [7:29:59<20:17:39, 134.30s/it]

(5175, 13)


 28%|█████████▋                         | 207/750 [7:32:12<20:09:29, 133.65s/it]

(5200, 13)


 28%|█████████▋                         | 208/750 [7:34:26<20:09:12, 133.86s/it]

(5225, 13)


 28%|█████████▊                         | 209/750 [7:36:40<20:06:13, 133.78s/it]

(5250, 13)


 28%|█████████▊                         | 210/750 [7:38:52<20:01:42, 133.52s/it]

(5275, 13)


 28%|█████████▊                         | 211/750 [7:41:06<19:59:24, 133.52s/it]

(5300, 13)


 28%|█████████▉                         | 212/750 [7:43:20<19:59:19, 133.75s/it]

(5325, 13)


 28%|█████████▉                         | 213/750 [7:45:35<20:00:26, 134.13s/it]

(5350, 13)


 29%|█████████▉                         | 214/750 [7:47:49<19:56:56, 133.99s/it]

(5375, 13)


 29%|██████████                         | 215/750 [7:50:05<20:00:27, 134.63s/it]

(5400, 13)


 29%|██████████                         | 216/750 [7:52:17<19:52:10, 133.95s/it]

(5425, 13)


 29%|██████████▏                        | 217/750 [7:54:30<19:45:37, 133.47s/it]

(5450, 13)


 29%|██████████▏                        | 218/750 [7:56:43<19:43:39, 133.49s/it]

(5475, 13)


 29%|██████████▏                        | 219/750 [7:58:58<19:44:59, 133.90s/it]

(5500, 13)


 29%|██████████▎                        | 220/750 [8:01:12<19:41:15, 133.73s/it]

(5525, 13)


 29%|██████████▎                        | 221/750 [8:03:23<19:34:26, 133.21s/it]

(5550, 13)


 30%|██████████▎                        | 222/750 [8:05:37<19:33:37, 133.37s/it]

(5575, 13)


 30%|██████████▍                        | 223/750 [8:07:49<19:28:14, 133.01s/it]

(5600, 13)


 30%|██████████▍                        | 224/750 [8:10:02<19:25:58, 133.00s/it]

(5625, 13)


 30%|██████████▌                        | 225/750 [8:12:21<19:38:05, 134.64s/it]

(5650, 13)


 30%|██████████▌                        | 226/750 [8:14:33<19:30:18, 134.00s/it]

(5675, 13)


 30%|██████████▌                        | 227/750 [8:16:47<19:26:47, 133.86s/it]

(5700, 13)


 30%|██████████▋                        | 228/750 [8:19:00<19:22:21, 133.60s/it]

(5725, 13)


 31%|██████████▋                        | 229/750 [8:21:13<19:18:36, 133.43s/it]

(5750, 13)


 31%|██████████▋                        | 230/750 [8:23:25<19:13:05, 133.05s/it]

(5775, 13)


 31%|██████████▊                        | 231/750 [8:25:38<19:10:25, 133.00s/it]

(5800, 13)


 31%|██████████▊                        | 232/750 [8:27:49<19:03:32, 132.46s/it]

(5825, 13)


 31%|██████████▊                        | 233/750 [8:30:02<19:01:08, 132.43s/it]

(5850, 13)


 31%|██████████▉                        | 234/750 [8:32:17<19:05:42, 133.22s/it]

(5875, 13)


 31%|██████████▉                        | 235/750 [8:34:29<19:01:47, 133.02s/it]

(5900, 13)


 31%|███████████                        | 236/750 [8:36:42<18:58:41, 132.92s/it]

(5925, 13)


 32%|███████████                        | 237/750 [8:38:55<18:57:33, 133.05s/it]

(5950, 13)


 32%|███████████                        | 238/750 [8:41:07<18:53:04, 132.78s/it]

(5975, 13)


 32%|███████████▏                       | 239/750 [8:43:22<18:55:20, 133.31s/it]

(6000, 13)


 32%|███████████▏                       | 240/750 [8:45:36<18:54:59, 133.53s/it]

(6025, 13)


 32%|███████████▏                       | 241/750 [8:47:51<18:56:31, 133.97s/it]

(6050, 13)


 32%|███████████▎                       | 242/750 [8:50:03<18:49:23, 133.39s/it]

(6075, 13)


 32%|███████████▎                       | 243/750 [8:52:18<18:51:29, 133.90s/it]

(6100, 13)


 33%|███████████▍                       | 244/750 [8:54:33<18:52:15, 134.26s/it]

(6125, 13)


 33%|███████████▍                       | 245/750 [8:56:49<18:54:12, 134.76s/it]

(6150, 13)


 33%|███████████▍                       | 246/750 [8:59:03<18:49:55, 134.52s/it]

(6175, 13)


 33%|███████████▌                       | 247/750 [9:01:17<18:46:00, 134.32s/it]

(6200, 13)


 33%|███████████▌                       | 248/750 [9:03:33<18:49:32, 135.00s/it]

(6225, 13)


 33%|███████████▌                       | 249/750 [9:05:48<18:45:06, 134.74s/it]

(6250, 13)


 33%|███████████▋                       | 250/750 [9:08:02<18:42:40, 134.72s/it]

(6275, 13)


 33%|███████████▋                       | 251/750 [9:10:17<18:41:09, 134.81s/it]

(6300, 13)


 34%|███████████▊                       | 252/750 [9:12:31<18:36:53, 134.56s/it]

(6325, 13)


 34%|███████████▊                       | 253/750 [9:14:44<18:29:45, 133.97s/it]

(6350, 13)


 34%|███████████▊                       | 254/750 [9:16:56<18:23:08, 133.45s/it]

(6375, 13)


 34%|███████████▉                       | 255/750 [9:19:11<18:24:43, 133.91s/it]

(6400, 13)


 34%|███████████▉                       | 256/750 [9:21:26<18:24:57, 134.21s/it]

(6425, 13)


 34%|███████████▉                       | 257/750 [9:23:40<18:21:16, 134.03s/it]

(6450, 13)


 34%|████████████                       | 258/750 [9:25:58<18:30:41, 135.45s/it]

(6475, 13)


 35%|████████████                       | 259/750 [9:28:15<18:30:20, 135.68s/it]

(6500, 13)


 35%|████████████▏                      | 260/750 [9:30:27<18:20:28, 134.75s/it]

(6525, 13)


 35%|████████████▏                      | 261/750 [9:32:41<18:16:23, 134.53s/it]

(6550, 13)


 35%|████████████▏                      | 262/750 [9:34:57<18:17:13, 134.90s/it]

(6575, 13)


 35%|████████████▎                      | 263/750 [9:37:11<18:12:51, 134.64s/it]

(6600, 13)


 35%|████████████▎                      | 264/750 [9:39:25<18:07:57, 134.32s/it]

(6625, 13)


 35%|████████████▎                      | 265/750 [9:41:38<18:03:53, 134.09s/it]

(6650, 13)


 35%|████████████▍                      | 266/750 [9:43:51<17:57:46, 133.61s/it]

(6675, 13)


 36%|████████████▍                      | 267/750 [9:46:04<17:54:09, 133.44s/it]

(6700, 13)


 36%|████████████▌                      | 268/750 [9:48:16<17:48:44, 133.04s/it]

(6725, 13)


 36%|████████████▌                      | 269/750 [9:50:29<17:48:07, 133.24s/it]

(6750, 13)


 36%|████████████▌                      | 270/750 [9:52:43<17:46:12, 133.28s/it]

(6775, 13)


 36%|████████████▋                      | 271/750 [9:54:58<17:48:16, 133.81s/it]

(6800, 13)


 36%|████████████▋                      | 272/750 [9:57:13<17:49:59, 134.31s/it]

(6825, 13)


 36%|████████████▋                      | 273/750 [9:59:26<17:44:56, 133.95s/it]

(6850, 13)


 37%|████████████▍                     | 274/750 [10:01:40<17:42:04, 133.88s/it]

(6875, 13)


 37%|████████████▍                     | 275/750 [10:03:52<17:34:03, 133.14s/it]

(6900, 13)


 37%|████████████▌                     | 276/750 [10:06:04<17:30:07, 132.93s/it]

(6925, 13)


 37%|████████████▌                     | 277/750 [10:08:17<17:27:30, 132.88s/it]

(6950, 13)


 37%|████████████▌                     | 278/750 [10:10:29<17:24:28, 132.77s/it]

(6975, 13)


 37%|████████████▋                     | 279/750 [10:12:51<17:43:04, 135.42s/it]

(7000, 13)


 37%|████████████▋                     | 280/750 [10:15:03<17:32:02, 134.30s/it]

(7025, 13)


 37%|████████████▋                     | 281/750 [10:17:15<17:25:38, 133.77s/it]

(7050, 13)


 38%|████████████▊                     | 282/750 [10:19:29<17:24:06, 133.86s/it]

(7075, 13)


 38%|████████████▊                     | 283/750 [10:21:44<17:23:51, 134.12s/it]

(7100, 13)


 38%|████████████▊                     | 284/750 [10:23:57<17:19:12, 133.80s/it]

(7125, 13)


 38%|████████████▉                     | 285/750 [10:26:11<17:16:19, 133.72s/it]

(7150, 13)


 38%|████████████▉                     | 286/750 [10:28:24<17:14:27, 133.77s/it]

(7175, 13)


 38%|█████████████                     | 287/750 [10:30:38<17:12:23, 133.79s/it]

(7200, 13)


 38%|█████████████                     | 288/750 [10:32:52<17:10:06, 133.78s/it]

(7225, 13)


 39%|█████████████                     | 289/750 [10:35:05<17:07:00, 133.67s/it]

(7250, 13)


 39%|█████████████▏                    | 290/750 [10:37:21<17:09:46, 134.32s/it]

(7275, 13)


 39%|█████████████▏                    | 291/750 [10:39:37<17:10:16, 134.68s/it]

(7300, 13)


 39%|█████████████▏                    | 292/750 [10:41:49<17:01:25, 133.81s/it]

(7325, 13)


 39%|█████████████▎                    | 293/750 [10:44:01<16:56:29, 133.46s/it]

(7350, 13)


 39%|█████████████▎                    | 294/750 [10:46:15<16:55:59, 133.68s/it]

(7375, 13)


 39%|█████████████▎                    | 295/750 [10:48:28<16:51:17, 133.36s/it]

(7400, 13)


 39%|█████████████▍                    | 296/750 [10:50:40<16:45:03, 132.83s/it]

(7425, 13)


 40%|█████████████▍                    | 297/750 [10:52:52<16:42:47, 132.82s/it]

(7450, 13)


 40%|█████████████▌                    | 298/750 [10:55:07<16:45:26, 133.47s/it]

(7475, 13)


 40%|█████████████▌                    | 299/750 [10:57:20<16:42:04, 133.31s/it]

(7500, 13)


 40%|█████████████▌                    | 300/750 [10:59:36<16:46:19, 134.18s/it]

(7525, 13)


 40%|█████████████▋                    | 301/750 [11:01:49<16:41:09, 133.79s/it]

(7550, 13)


 40%|█████████████▋                    | 302/750 [11:04:04<16:39:48, 133.90s/it]

(7575, 13)


 40%|█████████████▋                    | 303/750 [11:06:19<16:40:05, 134.24s/it]

(7600, 13)


 41%|█████████████▊                    | 304/750 [11:08:33<16:38:58, 134.39s/it]

(7625, 13)


 41%|█████████████▊                    | 305/750 [11:10:50<16:42:21, 135.15s/it]

(7625, 13)


 41%|██████████████▎                    | 306/750 [11:11:04<12:11:26, 98.84s/it]

(7625, 13)


 41%|██████████████▋                     | 307/750 [11:11:18<9:00:46, 73.24s/it]

(7650, 13)


 41%|██████████████▎                    | 308/750 [11:13:34<11:18:34, 92.11s/it]

(7675, 13)


 41%|██████████████                    | 309/750 [11:15:47<12:47:53, 104.47s/it]

(7700, 13)


 41%|██████████████                    | 310/750 [11:17:59<13:46:13, 112.67s/it]

(7725, 13)


 41%|██████████████                    | 311/750 [11:20:12<14:28:34, 118.71s/it]

(7750, 13)


 42%|██████████████▏                   | 312/750 [11:22:23<14:54:09, 122.49s/it]

(7775, 13)


 42%|██████████████▏                   | 313/750 [11:24:38<15:18:02, 126.05s/it]

(7800, 13)


 42%|██████████████▏                   | 314/750 [11:26:51<15:31:44, 128.22s/it]

(7825, 13)


 42%|██████████████▎                   | 315/750 [11:29:05<15:41:28, 129.86s/it]

(7850, 13)


 42%|██████████████▎                   | 316/750 [11:31:16<15:42:55, 130.36s/it]

(7875, 13)


 42%|██████████████▎                   | 317/750 [11:33:33<15:53:55, 132.18s/it]

(7900, 13)


 42%|██████████████▍                   | 318/750 [11:35:44<15:50:20, 131.99s/it]

(7925, 13)


 43%|██████████████▍                   | 319/750 [11:37:59<15:54:26, 132.87s/it]

(7950, 13)


 43%|██████████████▌                   | 320/750 [11:40:11<15:51:19, 132.74s/it]

(7975, 13)


 43%|██████████████▌                   | 321/750 [11:42:26<15:53:10, 133.31s/it]

(8000, 13)


 43%|██████████████▌                   | 322/750 [11:44:39<15:51:11, 133.35s/it]

(8025, 13)


 43%|██████████████▋                   | 323/750 [11:46:53<15:48:52, 133.33s/it]

(8050, 13)


 43%|██████████████▋                   | 324/750 [11:49:07<15:49:03, 133.67s/it]

(8075, 13)


 43%|██████████████▋                   | 325/750 [11:51:22<15:49:04, 133.99s/it]

(8100, 13)


 43%|██████████████▊                   | 326/750 [11:53:35<15:43:51, 133.56s/it]

(8125, 13)


 44%|██████████████▊                   | 327/750 [11:55:46<15:38:06, 133.07s/it]

(8150, 13)


 44%|██████████████▊                   | 328/750 [11:58:02<15:41:38, 133.88s/it]

(8175, 13)


 44%|██████████████▉                   | 329/750 [12:00:16<15:39:37, 133.91s/it]

(8200, 13)


 44%|██████████████▉                   | 330/750 [12:02:34<15:45:54, 135.13s/it]

(8225, 13)


 44%|███████████████                   | 331/750 [12:04:48<15:40:02, 134.61s/it]

(8250, 13)


 44%|███████████████                   | 332/750 [12:07:02<15:36:32, 134.43s/it]

(8275, 13)


 44%|███████████████                   | 333/750 [12:09:17<15:35:47, 134.65s/it]

(8300, 13)


 45%|███████████████▏                  | 334/750 [12:11:34<15:38:47, 135.40s/it]

(8325, 13)


 45%|███████████████▏                  | 335/750 [12:13:50<15:37:09, 135.49s/it]

(8350, 13)


 45%|███████████████▏                  | 336/750 [12:16:08<15:40:46, 136.34s/it]

(8375, 13)


 45%|███████████████▎                  | 337/750 [12:18:24<15:37:40, 136.22s/it]

(8400, 13)


 45%|███████████████▎                  | 338/750 [12:20:41<15:36:32, 136.39s/it]

(8425, 13)


 45%|███████████████▎                  | 339/750 [12:22:57<15:33:08, 136.23s/it]

(8450, 13)


 45%|███████████████▍                  | 340/750 [12:25:09<15:22:53, 135.06s/it]

(8475, 13)


 45%|███████████████▍                  | 341/750 [12:27:23<15:17:46, 134.64s/it]

(8500, 13)


 46%|███████████████▌                  | 342/750 [12:29:36<15:13:24, 134.33s/it]

(8525, 13)


 46%|███████████████▌                  | 343/750 [12:31:53<15:16:16, 135.08s/it]

(8550, 13)


 46%|███████████████▌                  | 344/750 [12:34:05<15:07:43, 134.15s/it]

(8575, 13)


 46%|███████████████▋                  | 345/750 [12:36:17<15:01:16, 133.52s/it]

(8600, 13)


 46%|███████████████▋                  | 346/750 [12:38:33<15:03:48, 134.23s/it]

(8625, 13)


 46%|███████████████▋                  | 347/750 [12:40:48<15:03:02, 134.45s/it]

(8650, 13)


 46%|███████████████▊                  | 348/750 [12:43:01<14:58:23, 134.09s/it]

(8675, 13)


 47%|███████████████▊                  | 349/750 [12:45:15<14:56:22, 134.12s/it]

(8700, 13)


 47%|███████████████▊                  | 350/750 [12:47:31<14:56:42, 134.51s/it]

(8725, 13)


 47%|███████████████▉                  | 351/750 [12:49:44<14:52:37, 134.23s/it]

(8750, 13)


 47%|███████████████▉                  | 352/750 [12:52:01<14:55:10, 134.95s/it]

(8775, 13)


 47%|████████████████                  | 353/750 [12:54:15<14:51:05, 134.67s/it]

(8800, 13)


 47%|████████████████                  | 354/750 [12:56:24<14:38:27, 133.10s/it]

(8825, 13)


 47%|████████████████                  | 355/750 [12:58:35<14:32:17, 132.50s/it]

(8850, 13)


 47%|████████████████▏                 | 356/750 [13:00:48<14:30:20, 132.54s/it]

(8875, 13)


 48%|████████████████▏                 | 357/750 [13:03:00<14:27:21, 132.42s/it]

(8900, 13)


 48%|████████████████▏                 | 358/750 [13:05:13<14:26:38, 132.65s/it]

(8925, 13)


 48%|████████████████▎                 | 359/750 [13:07:29<14:29:14, 133.39s/it]

(8950, 13)


 48%|████████████████▎                 | 360/750 [13:09:39<14:22:13, 132.65s/it]

(8975, 13)


 48%|████████████████▎                 | 361/750 [13:11:57<14:29:19, 134.09s/it]

(9000, 13)


 48%|████████████████▍                 | 362/750 [13:14:15<14:33:58, 135.15s/it]

(9025, 13)


 48%|████████████████▍                 | 363/750 [13:16:30<14:31:37, 135.14s/it]

(9050, 13)


 49%|████████████████▌                 | 364/750 [13:18:44<14:27:18, 134.81s/it]

(9075, 13)


 49%|████████████████▌                 | 365/750 [13:20:55<14:17:34, 133.65s/it]

(9100, 13)


 49%|████████████████▌                 | 366/750 [13:23:08<14:15:05, 133.61s/it]

(9125, 13)


 49%|████████████████▋                 | 367/750 [13:25:21<14:11:16, 133.36s/it]

(9150, 13)


 49%|████████████████▋                 | 368/750 [13:27:33<14:06:31, 132.96s/it]

(9175, 13)


 49%|████████████████▋                 | 369/750 [13:29:44<14:00:19, 132.33s/it]

(9200, 13)


 49%|████████████████▊                 | 370/750 [13:31:54<13:53:46, 131.65s/it]

(9225, 13)


 49%|████████████████▊                 | 371/750 [13:34:06<13:53:23, 131.93s/it]

(9250, 13)


 50%|████████████████▊                 | 372/750 [13:36:19<13:52:31, 132.15s/it]

(9275, 13)


 50%|████████████████▉                 | 373/750 [13:38:29<13:46:23, 131.52s/it]

(9300, 13)


 50%|████████████████▉                 | 374/750 [13:40:41<13:45:03, 131.66s/it]

(9325, 13)


 50%|█████████████████                 | 375/750 [13:42:51<13:39:45, 131.16s/it]

(9350, 13)


 50%|█████████████████                 | 376/750 [13:45:00<13:33:25, 130.50s/it]

(9375, 13)


 50%|█████████████████                 | 377/750 [13:47:12<13:34:48, 131.07s/it]

(9400, 13)


 50%|█████████████████▏                | 378/750 [13:49:26<13:37:40, 131.88s/it]

(9425, 13)


 51%|█████████████████▏                | 379/750 [13:51:41<13:40:16, 132.66s/it]

(9450, 13)


 51%|█████████████████▏                | 380/750 [13:53:55<13:41:23, 133.20s/it]

(9475, 13)


 51%|█████████████████▎                | 381/750 [13:56:11<13:43:13, 133.86s/it]

(9500, 13)


 51%|█████████████████▎                | 382/750 [13:58:23<13:38:24, 133.44s/it]

(9525, 13)


 51%|█████████████████▎                | 383/750 [14:00:37<13:36:34, 133.50s/it]

(9550, 13)


 51%|█████████████████▍                | 384/750 [14:02:49<13:32:08, 133.14s/it]

(9575, 13)


 51%|█████████████████▍                | 385/750 [14:05:04<13:32:33, 133.57s/it]

(9600, 13)


 51%|█████████████████▍                | 386/750 [14:07:19<13:34:16, 134.22s/it]

(9625, 13)


 52%|█████████████████▌                | 387/750 [14:09:32<13:28:34, 133.65s/it]

(9650, 13)


 52%|█████████████████▌                | 388/750 [14:11:52<13:39:12, 135.78s/it]

(9675, 13)


 52%|█████████████████▋                | 389/750 [14:14:06<13:32:31, 135.04s/it]

(9700, 13)


 52%|█████████████████▋                | 390/750 [14:16:19<13:26:56, 134.49s/it]

(9725, 13)


 52%|█████████████████▋                | 391/750 [14:18:31<13:20:03, 133.71s/it]

(9750, 13)


 52%|█████████████████▊                | 392/750 [14:20:47<13:22:11, 134.45s/it]

(9775, 13)


 52%|█████████████████▊                | 393/750 [14:23:00<13:17:02, 133.96s/it]

(9800, 13)


 53%|█████████████████▊                | 394/750 [14:25:12<13:12:15, 133.53s/it]

(9825, 13)


 53%|█████████████████▉                | 395/750 [14:27:26<13:10:35, 133.62s/it]

(9850, 13)


 53%|█████████████████▉                | 396/750 [14:29:42<13:12:22, 134.30s/it]

(9875, 13)


 53%|█████████████████▉                | 397/750 [14:31:57<13:11:07, 134.47s/it]

(9900, 13)


 53%|██████████████████                | 398/750 [14:34:14<13:12:47, 135.14s/it]

(9925, 13)


 53%|██████████████████                | 399/750 [14:36:27<13:08:13, 134.74s/it]

(9950, 13)


 53%|██████████████████▏               | 400/750 [14:38:46<13:12:04, 135.78s/it]

(9975, 13)


 53%|██████████████████▏               | 401/750 [14:40:58<13:03:24, 134.68s/it]

(10000, 13)


 54%|██████████████████▏               | 402/750 [14:43:10<12:57:44, 134.09s/it]

(10025, 13)


 54%|██████████████████▎               | 403/750 [14:45:25<12:56:46, 134.31s/it]

(10050, 13)


 54%|██████████████████▎               | 404/750 [14:47:40<12:55:51, 134.54s/it]

(10075, 13)


 54%|██████████████████▎               | 405/750 [14:49:53<12:49:35, 133.84s/it]

(10100, 13)


 54%|██████████████████▍               | 406/750 [14:52:05<12:45:05, 133.45s/it]

(10125, 13)


 54%|██████████████████▍               | 407/750 [14:54:21<12:46:49, 134.14s/it]

(10150, 13)


 54%|██████████████████▍               | 408/750 [14:56:34<12:42:44, 133.81s/it]

(10175, 13)


 55%|██████████████████▌               | 409/750 [14:58:51<12:46:42, 134.90s/it]

(10200, 13)


 55%|██████████████████▌               | 410/750 [15:01:07<12:46:25, 135.25s/it]

(10225, 13)


 55%|██████████████████▋               | 411/750 [15:03:23<12:45:28, 135.48s/it]

(10250, 13)


 55%|██████████████████▋               | 412/750 [15:05:38<12:41:39, 135.20s/it]

(10275, 13)


 55%|██████████████████▋               | 413/750 [15:07:52<12:37:42, 134.90s/it]

(10300, 13)


 55%|██████████████████▊               | 414/750 [15:10:10<12:40:31, 135.81s/it]

(10325, 13)


 55%|██████████████████▊               | 415/750 [15:12:40<13:02:26, 140.14s/it]

(10325, 13)


 55%|███████████████████▍               | 416/750 [15:12:54<9:28:14, 102.08s/it]

(10325, 13)


 56%|████████████████████                | 417/750 [15:13:08<7:00:25, 75.75s/it]

(10325, 13)


 56%|████████████████████                | 418/750 [15:13:21<5:15:43, 57.06s/it]

(10325, 13)


 56%|████████████████████                | 419/750 [15:13:35<4:02:41, 43.99s/it]

(10325, 13)


 56%|████████████████████▏               | 420/750 [15:13:48<3:11:50, 34.88s/it]

(10350, 13)


 56%|████████████████████▏               | 421/750 [15:16:16<6:15:51, 68.54s/it]

(10375, 13)


 56%|████████████████████▎               | 422/750 [15:18:39<8:16:47, 90.88s/it]

(10400, 13)


 56%|███████████████████▋               | 423/750 [15:20:58<9:33:53, 105.30s/it]

(10425, 13)


 57%|███████████████████▏              | 424/750 [15:23:19<10:31:47, 116.28s/it]

(10450, 13)


 57%|███████████████████▎              | 425/750 [15:25:36<11:03:12, 122.44s/it]

(10475, 13)


 57%|███████████████████▎              | 426/750 [15:27:57<11:31:15, 128.01s/it]

(10500, 13)


 57%|███████████████████▎              | 427/750 [15:30:16<11:46:39, 131.27s/it]

(10525, 13)


 57%|███████████████████▍              | 428/750 [15:32:35<11:56:25, 133.49s/it]

(10550, 13)


 57%|███████████████████▍              | 429/750 [15:34:52<12:00:51, 134.74s/it]

(10575, 13)


 57%|███████████████████▍              | 430/750 [15:37:12<12:06:10, 136.16s/it]

(10600, 13)


 57%|███████████████████▌              | 431/750 [15:39:27<12:02:38, 135.92s/it]

(10625, 13)


 58%|███████████████████▌              | 432/750 [15:41:48<12:08:23, 137.43s/it]

(10650, 13)


 58%|███████████████████▋              | 433/750 [15:44:08<12:09:19, 138.04s/it]

(10675, 13)


 58%|███████████████████▋              | 434/750 [15:46:25<12:05:51, 137.82s/it]

(10700, 13)


 58%|███████████████████▋              | 435/750 [15:48:37<11:54:24, 136.08s/it]

(10725, 13)


 58%|███████████████████▊              | 436/750 [15:50:47<11:42:22, 134.21s/it]

(10750, 13)


 58%|███████████████████▊              | 437/750 [15:52:58<11:34:41, 133.17s/it]

(10775, 13)


 58%|███████████████████▊              | 438/750 [15:55:08<11:27:58, 132.30s/it]

(10800, 13)


 59%|███████████████████▉              | 439/750 [15:57:25<11:33:23, 133.77s/it]

(10825, 13)


 59%|███████████████████▉              | 440/750 [15:59:35<11:25:08, 132.61s/it]

(10850, 13)


 59%|███████████████████▉              | 441/750 [16:01:50<11:27:19, 133.46s/it]

(10875, 13)


 59%|████████████████████              | 442/750 [16:04:01<11:20:39, 132.59s/it]

(10900, 13)


 59%|████████████████████              | 443/750 [16:06:18<11:25:27, 133.97s/it]

(10925, 13)


 59%|████████████████████▏             | 444/750 [16:08:27<11:15:05, 132.37s/it]

(10950, 13)


 59%|████████████████████▏             | 445/750 [16:10:37<11:09:53, 131.78s/it]

(10975, 13)


 59%|████████████████████▏             | 446/750 [16:12:49<11:07:07, 131.67s/it]

(11000, 13)


 60%|████████████████████▎             | 447/750 [16:14:59<11:03:08, 131.31s/it]

(11025, 13)


 60%|████████████████████▎             | 448/750 [16:17:09<10:58:53, 130.90s/it]

(11050, 13)


 60%|████████████████████▎             | 449/750 [16:19:21<10:57:30, 131.06s/it]

(11075, 13)


 60%|████████████████████▍             | 450/750 [16:21:48<11:20:12, 136.04s/it]

(11100, 13)


 60%|████████████████████▍             | 451/750 [16:23:59<11:10:34, 134.56s/it]

(11125, 13)


 60%|████████████████████▍             | 452/750 [16:26:09<11:01:25, 133.17s/it]

(11150, 13)


 60%|████████████████████▌             | 453/750 [16:28:20<10:56:05, 132.54s/it]

(11175, 13)


 61%|████████████████████▌             | 454/750 [16:30:32<10:53:11, 132.40s/it]

(11200, 13)


 61%|████████████████████▋             | 455/750 [16:32:41<10:45:55, 131.37s/it]

(11225, 13)


 61%|████████████████████▋             | 456/750 [16:34:52<10:42:38, 131.15s/it]

(11250, 13)


 61%|████████████████████▋             | 457/750 [16:37:07<10:46:36, 132.41s/it]

(11275, 13)


 61%|████████████████████▊             | 458/750 [16:39:18<10:41:52, 131.89s/it]

(11300, 13)


 61%|████████████████████▊             | 459/750 [16:41:27<10:35:58, 131.13s/it]

(11325, 13)


 61%|████████████████████▊             | 460/750 [16:43:43<10:40:09, 132.45s/it]

(11350, 13)


 61%|████████████████████▉             | 461/750 [16:45:54<10:36:20, 132.11s/it]

(11375, 13)


 62%|████████████████████▉             | 462/750 [16:48:06<10:33:18, 131.94s/it]

(11400, 13)


 62%|████████████████████▉             | 463/750 [16:50:16<10:28:20, 131.36s/it]

(11425, 13)


 62%|█████████████████████             | 464/750 [16:52:30<10:30:41, 132.31s/it]

(11450, 13)


 62%|█████████████████████             | 465/750 [16:54:41<10:26:51, 131.97s/it]

(11475, 13)


 62%|█████████████████████▏            | 466/750 [16:56:54<10:25:46, 132.21s/it]

(11500, 13)


 62%|█████████████████████▏            | 467/750 [16:59:06<10:22:28, 131.97s/it]

(11525, 13)


 62%|█████████████████████▏            | 468/750 [17:01:20<10:23:20, 132.62s/it]

(11550, 13)


 63%|█████████████████████▎            | 469/750 [17:03:32<10:20:08, 132.41s/it]

(11575, 13)


 63%|█████████████████████▎            | 470/750 [17:05:42<10:15:18, 131.85s/it]

(11600, 13)


 63%|█████████████████████▎            | 471/750 [17:07:59<10:19:36, 133.25s/it]

(11625, 13)


 63%|█████████████████████▍            | 472/750 [17:10:13<10:18:13, 133.43s/it]

(11650, 13)


 63%|█████████████████████▍            | 473/750 [17:12:24<10:13:31, 132.89s/it]

(11675, 13)


 63%|█████████████████████▍            | 474/750 [17:14:39<10:14:31, 133.59s/it]

(11700, 13)


 63%|█████████████████████▌            | 475/750 [17:16:53<10:12:22, 133.61s/it]

(11725, 13)


 63%|█████████████████████▌            | 476/750 [17:19:07<10:10:19, 133.65s/it]

(11750, 13)


 64%|█████████████████████▌            | 477/750 [17:21:21<10:08:49, 133.81s/it]

(11775, 13)


 64%|█████████████████████▋            | 478/750 [17:23:34<10:05:23, 133.54s/it]

(11800, 13)


 64%|█████████████████████▋            | 479/750 [17:25:50<10:06:07, 134.20s/it]

(11825, 13)


 64%|█████████████████████▊            | 480/750 [17:28:05<10:05:05, 134.46s/it]

(11850, 13)


 64%|██████████████████████▍            | 481/750 [17:30:14<9:56:27, 133.04s/it]

(11875, 13)


 64%|██████████████████████▍            | 482/750 [17:32:26<9:51:32, 132.43s/it]

(11900, 13)


 64%|██████████████████████▌            | 483/750 [17:34:36<9:46:10, 131.73s/it]

(11925, 13)


 65%|██████████████████████▌            | 484/750 [17:36:47<9:42:57, 131.49s/it]

(11950, 13)


 65%|██████████████████████▋            | 485/750 [17:39:00<9:43:23, 132.09s/it]

(11975, 13)


 65%|██████████████████████▋            | 486/750 [17:41:10<9:38:53, 131.56s/it]

(12000, 13)


 65%|██████████████████████▋            | 487/750 [17:43:22<9:36:08, 131.44s/it]

(12025, 13)


 65%|██████████████████████▊            | 488/750 [17:45:33<9:33:44, 131.39s/it]

(12050, 13)


 65%|██████████████████████▊            | 489/750 [17:47:47<9:34:53, 132.16s/it]

(12075, 13)


 65%|██████████████████████▊            | 490/750 [17:49:56<9:28:32, 131.20s/it]

(12100, 13)


 65%|██████████████████████▉            | 491/750 [17:52:05<9:23:41, 130.58s/it]

(12125, 13)


 66%|██████████████████████▉            | 492/750 [17:54:17<9:23:17, 131.00s/it]

(12150, 13)


 66%|███████████████████████            | 493/750 [17:56:29<9:22:31, 131.33s/it]

(12175, 13)


 66%|███████████████████████            | 494/750 [17:58:40<9:20:31, 131.37s/it]

(12200, 13)


 66%|███████████████████████            | 495/750 [18:00:50<9:15:46, 130.77s/it]

(12225, 13)


 66%|███████████████████████▏           | 496/750 [18:03:00<9:12:54, 130.61s/it]

(12250, 13)


 66%|███████████████████████▏           | 497/750 [18:05:08<9:07:51, 129.93s/it]

(12275, 13)


 66%|███████████████████████▏           | 498/750 [18:07:28<9:17:42, 132.79s/it]

(12300, 13)


 67%|███████████████████████▎           | 499/750 [18:09:40<9:14:18, 132.50s/it]

(12325, 13)


 67%|███████████████████████▎           | 500/750 [18:11:52<9:11:40, 132.40s/it]

(12350, 13)


 67%|███████████████████████▍           | 501/750 [18:14:04<9:09:33, 132.43s/it]

(12375, 13)


 67%|███████████████████████▍           | 502/750 [18:16:19<9:09:51, 133.03s/it]

(12400, 13)


 67%|███████████████████████▍           | 503/750 [18:18:30<9:05:13, 132.44s/it]

(12425, 13)


 67%|███████████████████████▌           | 504/750 [18:20:45<9:06:13, 133.23s/it]

(12450, 13)


 67%|███████████████████████▌           | 505/750 [18:22:54<8:59:19, 132.08s/it]

(12475, 13)


 67%|███████████████████████▌           | 506/750 [18:25:04<8:54:35, 131.46s/it]

(12500, 13)


 68%|███████████████████████▋           | 507/750 [18:27:12<8:48:10, 130.41s/it]

(12525, 13)


 68%|███████████████████████▋           | 508/750 [18:29:21<8:44:29, 130.04s/it]

(12550, 13)


 68%|███████████████████████▊           | 509/750 [18:31:29<8:39:51, 129.42s/it]

(12575, 13)


 68%|███████████████████████▊           | 510/750 [18:33:39<8:37:47, 129.45s/it]

(12600, 13)


 68%|███████████████████████▊           | 511/750 [18:35:50<8:37:14, 129.85s/it]

(12625, 13)


 68%|███████████████████████▉           | 512/750 [18:38:06<8:43:15, 131.91s/it]

(12650, 13)


 68%|███████████████████████▉           | 513/750 [18:40:17<8:39:30, 131.52s/it]

(12675, 13)


 69%|███████████████████████▉           | 514/750 [18:42:29<8:38:13, 131.75s/it]

(12700, 13)


 69%|████████████████████████           | 515/750 [18:44:45<8:40:25, 132.87s/it]

(12725, 13)


 69%|████████████████████████           | 516/750 [18:46:58<8:38:44, 133.01s/it]

(12750, 13)


 69%|████████████████████████▏          | 517/750 [18:49:20<8:46:57, 135.70s/it]

(12775, 13)


 69%|████████████████████████▏          | 518/750 [18:51:37<8:45:43, 135.97s/it]

(12800, 13)


 69%|████████████████████████▏          | 519/750 [18:53:50<8:40:34, 135.21s/it]

(12825, 13)


 69%|████████████████████████▎          | 520/750 [18:56:02<8:34:46, 134.29s/it]

(12850, 13)


 69%|████████████████████████▎          | 521/750 [18:58:13<8:28:58, 133.36s/it]

(12875, 13)


 70%|████████████████████████▎          | 522/750 [19:00:27<8:27:14, 133.49s/it]

(12900, 13)


 70%|████████████████████████▍          | 523/750 [19:02:40<8:23:53, 133.19s/it]

(12925, 13)


 70%|████████████████████████▍          | 524/750 [19:04:56<8:25:42, 134.26s/it]

(12950, 13)


 70%|████████████████████████▌          | 525/750 [19:07:10<8:22:35, 134.03s/it]

(12975, 13)


 70%|████████████████████████▌          | 526/750 [19:09:24<8:20:05, 133.95s/it]

(13000, 13)


 70%|████████████████████████▌          | 527/750 [19:11:35<8:15:16, 133.26s/it]

(13025, 13)


 70%|████████████████████████▋          | 528/750 [19:13:46<8:10:27, 132.55s/it]

(13050, 13)


 71%|████████████████████████▋          | 529/750 [19:15:59<8:08:35, 132.65s/it]

(13075, 13)


 71%|████████████████████████▋          | 530/750 [19:18:12<8:06:22, 132.65s/it]

(13100, 13)


 71%|████████████████████████▊          | 531/750 [19:20:22<8:01:57, 132.04s/it]

(13125, 13)


 71%|████████████████████████▊          | 532/750 [19:22:38<8:03:06, 132.97s/it]

(13150, 13)


 71%|████████████████████████▊          | 533/750 [19:24:47<7:57:06, 131.92s/it]

(13175, 13)


 71%|████████████████████████▉          | 534/750 [19:26:58<7:53:33, 131.54s/it]

(13200, 13)


 71%|████████████████████████▉          | 535/750 [19:29:13<7:55:51, 132.80s/it]

(13225, 13)


 71%|█████████████████████████          | 536/750 [19:31:22<7:49:27, 131.62s/it]

(13250, 13)


 72%|█████████████████████████          | 537/750 [19:33:33<7:46:43, 131.47s/it]

(13275, 13)


 72%|█████████████████████████          | 538/750 [19:35:44<7:43:42, 131.24s/it]

(13300, 13)


 72%|█████████████████████████▏         | 539/750 [19:38:00<7:46:17, 132.60s/it]

(13325, 13)


 72%|█████████████████████████▏         | 540/750 [19:40:09<7:40:33, 131.59s/it]

(13350, 13)


 72%|█████████████████████████▏         | 541/750 [19:42:18<7:35:11, 130.68s/it]

(13375, 13)


 72%|█████████████████████████▎         | 542/750 [19:44:31<7:35:59, 131.54s/it]

(13400, 13)


 72%|█████████████████████████▎         | 543/750 [19:46:41<7:31:38, 130.91s/it]

(13425, 13)


 73%|█████████████████████████▍         | 544/750 [19:48:52<7:29:29, 130.92s/it]

(13450, 13)


 73%|█████████████████████████▍         | 545/750 [19:51:02<7:26:36, 130.71s/it]

(13475, 13)


 73%|█████████████████████████▍         | 546/750 [19:53:10<7:21:28, 129.84s/it]

(13500, 13)


 73%|█████████████████████████▌         | 547/750 [19:55:20<7:19:48, 130.00s/it]

(13525, 13)


 73%|█████████████████████████▌         | 548/750 [19:57:30<7:17:25, 129.93s/it]

(13550, 13)


 73%|█████████████████████████▌         | 549/750 [19:59:40<7:15:55, 130.13s/it]

(13575, 13)


 73%|█████████████████████████▋         | 550/750 [20:01:54<7:17:45, 131.33s/it]

(13600, 13)


 73%|█████████████████████████▋         | 551/750 [20:04:08<7:18:06, 132.09s/it]

(13625, 13)


 74%|█████████████████████████▊         | 552/750 [20:06:22<7:17:14, 132.50s/it]

(13650, 13)


 74%|█████████████████████████▊         | 553/750 [20:08:33<7:14:14, 132.25s/it]

(13675, 13)


 74%|█████████████████████████▊         | 554/750 [20:10:45<7:11:09, 131.99s/it]

(13700, 13)


 74%|█████████████████████████▉         | 555/750 [20:12:57<7:09:07, 132.04s/it]

(13725, 13)


 74%|█████████████████████████▉         | 556/750 [20:15:21<7:18:13, 135.53s/it]

(13750, 13)


 74%|█████████████████████████▉         | 557/750 [20:17:35<7:15:13, 135.30s/it]

(13775, 13)


 74%|██████████████████████████         | 558/750 [20:19:49<7:11:41, 134.90s/it]

(13800, 13)


 75%|██████████████████████████         | 559/750 [20:22:06<7:11:07, 135.43s/it]

(13825, 13)


 75%|██████████████████████████▏        | 560/750 [20:24:22<7:08:56, 135.45s/it]

(13850, 13)


 75%|██████████████████████████▏        | 561/750 [20:26:34<7:04:00, 134.61s/it]

(13875, 13)


 75%|██████████████████████████▏        | 562/750 [20:28:50<7:02:54, 134.97s/it]

(13900, 13)


 75%|██████████████████████████▎        | 563/750 [20:31:06<7:01:15, 135.17s/it]

(13925, 13)


 75%|██████████████████████████▎        | 564/750 [20:33:22<7:00:00, 135.49s/it]

(13950, 13)


 75%|██████████████████████████▎        | 565/750 [20:35:36<6:56:49, 135.18s/it]

(13975, 13)


 75%|██████████████████████████▍        | 566/750 [20:37:53<6:55:44, 135.57s/it]

(14000, 13)


 76%|██████████████████████████▍        | 567/750 [20:40:06<6:51:41, 134.98s/it]

(14025, 13)


 76%|██████████████████████████▌        | 568/750 [20:42:20<6:48:24, 134.64s/it]

(14050, 13)


 76%|██████████████████████████▌        | 569/750 [20:44:36<6:46:46, 134.84s/it]

(14075, 13)


 76%|██████████████████████████▌        | 570/750 [20:46:51<6:44:49, 134.94s/it]

(14100, 13)


 76%|██████████████████████████▋        | 571/750 [20:49:08<6:44:51, 135.71s/it]

(14125, 13)


 76%|██████████████████████████▋        | 572/750 [20:51:21<6:39:56, 134.81s/it]

(14150, 13)


 76%|██████████████████████████▋        | 573/750 [20:53:36<6:37:54, 134.88s/it]

(14175, 13)


 77%|██████████████████████████▊        | 574/750 [20:55:51<6:36:01, 135.01s/it]

(14200, 13)


 77%|██████████████████████████▊        | 575/750 [20:58:10<6:37:15, 136.20s/it]

(14225, 13)


 77%|██████████████████████████▉        | 576/750 [21:00:28<6:35:59, 136.55s/it]

(14250, 13)


 77%|██████████████████████████▉        | 577/750 [21:02:41<6:30:53, 135.57s/it]

(14275, 13)


 77%|██████████████████████████▉        | 578/750 [21:04:59<6:30:20, 136.17s/it]

(14300, 13)


 77%|███████████████████████████        | 579/750 [21:07:15<6:28:31, 136.33s/it]

(14325, 13)


 77%|███████████████████████████        | 580/750 [21:09:33<6:27:32, 136.78s/it]

(14350, 13)


 77%|███████████████████████████        | 581/750 [21:11:47<6:22:26, 135.78s/it]

(14375, 13)


 78%|███████████████████████████▏       | 582/750 [21:14:05<6:22:47, 136.71s/it]

(14400, 13)


 78%|███████████████████████████▏       | 583/750 [21:16:23<6:21:14, 136.97s/it]

(14425, 13)


 78%|███████████████████████████▎       | 584/750 [21:18:43<6:21:14, 137.80s/it]

(14450, 13)


 78%|███████████████████████████▎       | 585/750 [21:20:54<6:13:30, 135.82s/it]

(14475, 13)


 78%|███████████████████████████▎       | 586/750 [21:23:14<6:14:19, 136.95s/it]

(14500, 13)


 78%|███████████████████████████▍       | 587/750 [21:26:15<6:48:28, 150.36s/it]

(14525, 13)


 78%|███████████████████████████▍       | 588/750 [21:28:30<6:33:45, 145.84s/it]

(14550, 13)


 79%|███████████████████████████▍       | 589/750 [21:30:45<6:22:27, 142.53s/it]

(14575, 13)


 79%|███████████████████████████▌       | 590/750 [21:33:00<6:13:36, 140.10s/it]

(14600, 13)


 79%|███████████████████████████▌       | 591/750 [21:35:14<6:06:42, 138.38s/it]

(14625, 13)


 79%|███████████████████████████▋       | 592/750 [21:37:33<6:04:39, 138.48s/it]

(14650, 13)


 79%|███████████████████████████▋       | 593/750 [21:39:47<5:59:00, 137.20s/it]

(14675, 13)


 79%|███████████████████████████▋       | 594/750 [21:42:06<5:58:25, 137.86s/it]

(14700, 13)


 79%|███████████████████████████▊       | 595/750 [21:44:35<6:04:25, 141.07s/it]

(14725, 13)


 79%|███████████████████████████▊       | 596/750 [21:46:58<6:03:30, 141.63s/it]

(14750, 13)


 80%|███████████████████████████▊       | 597/750 [21:49:26<6:06:14, 143.62s/it]

(14775, 13)


 80%|███████████████████████████▉       | 598/750 [21:51:45<6:00:05, 142.14s/it]

(14800, 13)


 80%|███████████████████████████▉       | 599/750 [21:54:01<5:53:17, 140.38s/it]

(14825, 13)


 80%|████████████████████████████       | 600/750 [21:56:17<5:47:41, 139.08s/it]

(14850, 13)


 80%|████████████████████████████       | 601/750 [21:58:30<5:40:54, 137.28s/it]

(14875, 13)


 80%|████████████████████████████       | 602/750 [22:00:47<5:37:54, 136.99s/it]

(14900, 13)


 80%|████████████████████████████▏      | 603/750 [22:03:04<5:35:46, 137.05s/it]

(14925, 13)


 81%|████████████████████████████▏      | 604/750 [22:05:23<5:35:04, 137.70s/it]

(14950, 13)


 81%|████████████████████████████▏      | 605/750 [22:07:43<5:34:10, 138.28s/it]

(14975, 13)


 81%|████████████████████████████▎      | 606/750 [22:09:56<5:28:40, 136.95s/it]

(15000, 13)


 81%|████████████████████████████▎      | 607/750 [22:12:09<5:23:13, 135.62s/it]

(15025, 13)


 81%|████████████████████████████▎      | 608/750 [22:14:24<5:20:42, 135.51s/it]

(15050, 13)


 81%|████████████████████████████▍      | 609/750 [22:16:39<5:17:48, 135.24s/it]

(15075, 13)


 81%|████████████████████████████▍      | 610/750 [22:18:54<5:15:48, 135.35s/it]

(15100, 13)


 81%|████████████████████████████▌      | 611/750 [22:21:08<5:12:11, 134.76s/it]

(15125, 13)


 82%|████████████████████████████▌      | 612/750 [22:23:22<5:09:43, 134.67s/it]

(15150, 13)


 82%|████████████████████████████▌      | 613/750 [22:25:37<5:07:37, 134.73s/it]

(15175, 13)


 82%|████████████████████████████▋      | 614/750 [22:27:51<5:04:58, 134.55s/it]

(15200, 13)


 82%|████████████████████████████▋      | 615/750 [22:30:07<5:03:36, 134.94s/it]

(15225, 13)


 82%|████████████████████████████▋      | 616/750 [22:32:28<5:05:17, 136.70s/it]

(15250, 13)


 82%|████████████████████████████▊      | 617/750 [22:34:43<5:02:16, 136.37s/it]

(15275, 13)


 82%|████████████████████████████▊      | 618/750 [22:37:03<5:02:21, 137.44s/it]

(15300, 13)


 83%|████████████████████████████▉      | 619/750 [22:39:20<4:59:16, 137.07s/it]

(15325, 13)


 83%|████████████████████████████▉      | 620/750 [22:41:35<4:56:07, 136.67s/it]

(15350, 13)


 83%|████████████████████████████▉      | 621/750 [22:43:52<4:54:05, 136.79s/it]

(15375, 13)


 83%|█████████████████████████████      | 622/750 [22:46:12<4:53:23, 137.53s/it]

(15400, 13)


 83%|█████████████████████████████      | 623/750 [22:48:30<4:51:42, 137.82s/it]

(15425, 13)


 83%|█████████████████████████████      | 624/750 [22:50:49<4:50:04, 138.13s/it]

(15450, 13)


 83%|█████████████████████████████▏     | 625/750 [22:53:05<4:46:39, 137.60s/it]

(15475, 13)


 83%|█████████████████████████████▏     | 626/750 [22:55:22<4:43:34, 137.22s/it]

(15500, 13)


 84%|█████████████████████████████▎     | 627/750 [22:57:42<4:43:17, 138.19s/it]

(15525, 13)


 84%|█████████████████████████████▎     | 628/750 [22:59:59<4:40:20, 137.87s/it]

(15550, 13)


 84%|█████████████████████████████▎     | 629/750 [23:02:16<4:37:13, 137.47s/it]

(15575, 13)


 84%|█████████████████████████████▍     | 630/750 [23:04:33<4:34:32, 137.27s/it]

(15600, 13)


 84%|█████████████████████████████▍     | 631/750 [23:06:48<4:31:21, 136.82s/it]

(15625, 13)


 84%|█████████████████████████████▍     | 632/750 [23:09:08<4:30:30, 137.55s/it]

(15650, 13)


 84%|█████████████████████████████▌     | 633/750 [23:11:26<4:28:27, 137.67s/it]

(15675, 13)


 85%|█████████████████████████████▌     | 634/750 [23:13:45<4:27:20, 138.28s/it]

(15700, 13)


 85%|█████████████████████████████▋     | 635/750 [23:16:03<4:24:49, 138.17s/it]

(15725, 13)


 85%|█████████████████████████████▋     | 636/750 [23:18:19<4:21:20, 137.55s/it]

(15750, 13)


 85%|█████████████████████████████▋     | 637/750 [23:20:38<4:19:23, 137.73s/it]

(15775, 13)


 85%|█████████████████████████████▊     | 638/750 [23:22:57<4:18:10, 138.30s/it]

(15800, 13)


 85%|█████████████████████████████▊     | 639/750 [23:25:13<4:14:41, 137.67s/it]

(15825, 13)


 85%|█████████████████████████████▊     | 640/750 [23:27:31<4:12:37, 137.79s/it]

(15850, 13)


 85%|█████████████████████████████▉     | 641/750 [23:29:50<4:11:00, 138.17s/it]

(15875, 13)


 86%|█████████████████████████████▉     | 642/750 [23:32:14<4:11:47, 139.89s/it]

(15900, 13)


 86%|██████████████████████████████     | 643/750 [23:34:32<4:08:06, 139.12s/it]

(15925, 13)


 86%|██████████████████████████████     | 644/750 [23:36:50<4:05:14, 138.82s/it]

(15950, 13)


 86%|██████████████████████████████     | 645/750 [23:39:11<4:04:18, 139.60s/it]

(15975, 13)


 86%|██████████████████████████████▏    | 646/750 [23:41:27<3:59:52, 138.39s/it]

(16000, 13)


 86%|██████████████████████████████▏    | 647/750 [23:43:56<4:03:13, 141.68s/it]

(16025, 13)


 86%|██████████████████████████████▏    | 648/750 [23:46:28<4:06:15, 144.86s/it]

(16050, 13)


 87%|██████████████████████████████▎    | 649/750 [23:48:51<4:02:31, 144.07s/it]

(16075, 13)


 87%|██████████████████████████████▎    | 650/750 [23:51:11<3:58:21, 143.02s/it]

(16100, 13)


 87%|██████████████████████████████▍    | 651/750 [23:53:28<3:52:57, 141.18s/it]

(16125, 13)


 87%|██████████████████████████████▍    | 652/750 [23:55:48<3:50:11, 140.93s/it]

(16150, 13)


 87%|██████████████████████████████▍    | 653/750 [23:58:06<3:46:13, 139.93s/it]

(16175, 13)


 87%|██████████████████████████████▌    | 654/750 [24:00:27<3:44:19, 140.20s/it]

(16200, 13)


 87%|██████████████████████████████▌    | 655/750 [24:02:45<3:41:05, 139.64s/it]

(16225, 13)


 87%|██████████████████████████████▌    | 656/750 [24:05:02<3:37:28, 138.81s/it]

(16250, 13)


 88%|██████████████████████████████▋    | 657/750 [24:07:20<3:34:52, 138.63s/it]

(16275, 13)


 88%|██████████████████████████████▋    | 658/750 [24:09:38<3:32:03, 138.30s/it]

(16300, 13)


 88%|██████████████████████████████▊    | 659/750 [24:11:58<3:30:24, 138.73s/it]

(16325, 13)


 88%|██████████████████████████████▊    | 660/750 [24:14:15<3:27:37, 138.41s/it]

(16350, 13)


 88%|██████████████████████████████▊    | 661/750 [24:16:35<3:25:54, 138.82s/it]

(16375, 13)


 88%|██████████████████████████████▉    | 662/750 [24:18:49<3:21:34, 137.43s/it]

(16400, 13)


 88%|██████████████████████████████▉    | 663/750 [24:21:05<3:18:31, 136.91s/it]

(16425, 13)


 89%|██████████████████████████████▉    | 664/750 [24:23:19<3:14:51, 135.95s/it]

(16450, 13)


 89%|███████████████████████████████    | 665/750 [24:25:33<3:11:45, 135.36s/it]

(16475, 13)


 89%|███████████████████████████████    | 666/750 [24:27:51<3:10:51, 136.33s/it]

(16500, 13)


 89%|███████████████████████████████▏   | 667/750 [24:30:08<3:08:55, 136.57s/it]

(16525, 13)


 89%|███████████████████████████████▏   | 668/750 [24:32:28<3:07:56, 137.52s/it]

(16550, 13)


 89%|███████████████████████████████▏   | 669/750 [24:34:47<3:06:17, 137.99s/it]

(16575, 13)


 89%|███████████████████████████████▎   | 670/750 [24:37:04<3:03:33, 137.67s/it]

(16600, 13)


 89%|███████████████████████████████▎   | 671/750 [24:39:20<3:00:45, 137.28s/it]

(16625, 13)


 90%|███████████████████████████████▎   | 672/750 [24:41:41<2:59:50, 138.34s/it]

(16650, 13)


 90%|███████████████████████████████▍   | 673/750 [24:43:58<2:56:52, 137.83s/it]

(16675, 13)


 90%|███████████████████████████████▍   | 674/750 [24:46:16<2:54:36, 137.85s/it]

(16700, 13)


 90%|███████████████████████████████▌   | 675/750 [24:48:33<2:52:06, 137.68s/it]

(16725, 13)


 90%|███████████████████████████████▌   | 676/750 [24:50:52<2:50:10, 137.98s/it]

(16750, 13)


 90%|███████████████████████████████▌   | 677/750 [24:53:07<2:46:49, 137.11s/it]

(16775, 13)


 90%|███████████████████████████████▋   | 678/750 [24:55:23<2:44:05, 136.74s/it]

(16800, 13)


 91%|███████████████████████████████▋   | 679/750 [24:57:40<2:42:09, 137.04s/it]

(16825, 13)


 91%|███████████████████████████████▋   | 680/750 [24:59:57<2:39:40, 136.86s/it]

(16850, 13)


 91%|███████████████████████████████▊   | 681/750 [25:02:18<2:38:53, 138.16s/it]

(16875, 13)


 91%|███████████████████████████████▊   | 682/750 [25:04:37<2:36:54, 138.44s/it]

(16900, 13)


 91%|███████████████████████████████▊   | 683/750 [25:06:54<2:34:11, 138.09s/it]

(16925, 13)


 91%|███████████████████████████████▉   | 684/750 [25:09:13<2:31:55, 138.11s/it]

(16950, 13)


 91%|███████████████████████████████▉   | 685/750 [25:11:28<2:28:48, 137.37s/it]

(16975, 13)


 91%|████████████████████████████████   | 686/750 [25:13:47<2:27:00, 137.82s/it]

(17000, 13)


 92%|████████████████████████████████   | 687/750 [25:16:07<2:25:20, 138.42s/it]

(17025, 13)


 92%|████████████████████████████████   | 688/750 [25:18:26<2:23:09, 138.54s/it]

(17050, 13)


 92%|████████████████████████████████▏  | 689/750 [25:20:46<2:21:23, 139.07s/it]

(17075, 13)


 92%|████████████████████████████████▏  | 690/750 [25:23:06<2:19:13, 139.23s/it]

(17100, 13)


 92%|████████████████████████████████▏  | 691/750 [25:25:27<2:17:33, 139.89s/it]

(17125, 13)


 92%|████████████████████████████████▎  | 692/750 [25:27:46<2:14:55, 139.58s/it]

(17150, 13)


 92%|████████████████████████████████▎  | 693/750 [25:30:04<2:12:04, 139.03s/it]

(17175, 13)


 93%|████████████████████████████████▍  | 694/750 [25:32:23<2:09:43, 138.98s/it]

(17200, 13)


 93%|████████████████████████████████▍  | 695/750 [25:34:39<2:06:45, 138.28s/it]

(17225, 13)


 93%|████████████████████████████████▍  | 696/750 [25:36:55<2:03:41, 137.43s/it]

(17250, 13)


 93%|████████████████████████████████▌  | 697/750 [25:39:14<2:01:47, 137.88s/it]

(17275, 13)


 93%|████████████████████████████████▌  | 698/750 [25:41:31<1:59:21, 137.72s/it]

(17300, 13)


 93%|████████████████████████████████▌  | 699/750 [25:43:50<1:57:19, 138.04s/it]

(17325, 13)


 93%|████████████████████████████████▋  | 700/750 [25:46:12<1:56:01, 139.24s/it]

(17350, 13)


 93%|████████████████████████████████▋  | 701/750 [25:48:28<1:53:00, 138.38s/it]

(17375, 13)


 94%|████████████████████████████████▊  | 702/750 [25:50:45<1:50:24, 138.02s/it]

(17400, 13)


 94%|████████████████████████████████▊  | 703/750 [25:53:03<1:48:01, 137.91s/it]

(17425, 13)


 94%|████████████████████████████████▊  | 704/750 [25:55:22<1:46:04, 138.37s/it]

(17450, 13)


 94%|████████████████████████████████▉  | 705/750 [25:57:47<1:45:06, 140.15s/it]

(17475, 13)


 94%|████████████████████████████████▉  | 706/750 [26:00:07<1:42:46, 140.16s/it]

(17500, 13)


 94%|████████████████████████████████▉  | 707/750 [26:02:27<1:40:26, 140.16s/it]

(17525, 13)


 94%|█████████████████████████████████  | 708/750 [26:04:45<1:37:40, 139.53s/it]

(17550, 13)


 95%|█████████████████████████████████  | 709/750 [26:07:02<1:34:52, 138.83s/it]

(17575, 13)


 95%|█████████████████████████████████▏ | 710/750 [26:09:20<1:32:17, 138.45s/it]

(17600, 13)


 95%|█████████████████████████████████▏ | 711/750 [26:11:39<1:30:08, 138.67s/it]

(17625, 13)


 95%|█████████████████████████████████▏ | 712/750 [26:14:00<1:28:18, 139.44s/it]

(17650, 13)


 95%|█████████████████████████████████▎ | 713/750 [26:16:19<1:25:49, 139.17s/it]

(17675, 13)


 95%|█████████████████████████████████▎ | 714/750 [26:18:37<1:23:19, 138.88s/it]

(17700, 13)


 95%|█████████████████████████████████▎ | 715/750 [26:21:06<1:22:47, 141.92s/it]

(17725, 13)


 95%|█████████████████████████████████▍ | 716/750 [26:23:26<1:20:01, 141.21s/it]

(17750, 13)


 96%|█████████████████████████████████▍ | 717/750 [26:25:45<1:17:25, 140.77s/it]

(17775, 13)


 96%|█████████████████████████████████▌ | 718/750 [26:28:05<1:14:51, 140.35s/it]

(17800, 13)


 96%|█████████████████████████████████▌ | 719/750 [26:30:24<1:12:19, 139.97s/it]

(17825, 13)


 96%|█████████████████████████████████▌ | 720/750 [26:32:43<1:09:53, 139.77s/it]

(17850, 13)


 96%|█████████████████████████████████▋ | 721/750 [26:35:00<1:07:07, 138.88s/it]

(17875, 13)


 96%|█████████████████████████████████▋ | 722/750 [26:37:16<1:04:27, 138.14s/it]

(17900, 13)


 96%|█████████████████████████████████▋ | 723/750 [26:39:36<1:02:19, 138.48s/it]

(17925, 13)


 97%|█████████████████████████████████▊ | 724/750 [26:41:54<1:00:03, 138.60s/it]

(17950, 13)


 97%|███████████████████████████████████▊ | 725/750 [26:44:15<57:59, 139.17s/it]

(17975, 13)


 97%|███████████████████████████████████▊ | 726/750 [26:46:38<56:04, 140.21s/it]

(18000, 13)


 97%|███████████████████████████████████▊ | 727/750 [26:48:57<53:42, 140.10s/it]

(18025, 13)


 97%|███████████████████████████████████▉ | 728/750 [26:51:20<51:37, 140.80s/it]

(18050, 13)


 97%|███████████████████████████████████▉ | 729/750 [26:53:40<49:14, 140.68s/it]

(18075, 13)


 97%|████████████████████████████████████ | 730/750 [26:56:02<46:57, 140.87s/it]

(18100, 13)


 97%|████████████████████████████████████ | 731/750 [26:58:24<44:47, 141.45s/it]

(18125, 13)


 98%|████████████████████████████████████ | 732/750 [27:00:54<43:08, 143.82s/it]

(18150, 13)


 98%|████████████████████████████████████▏| 733/750 [27:03:18<40:48, 144.02s/it]

(18175, 13)


 98%|████████████████████████████████████▏| 734/750 [27:05:41<38:20, 143.76s/it]

(18200, 13)


 98%|████████████████████████████████████▎| 735/750 [27:08:06<35:59, 143.95s/it]

(18225, 13)


 98%|████████████████████████████████████▎| 736/750 [27:10:29<33:33, 143.83s/it]

(18250, 13)


 98%|████████████████████████████████████▎| 737/750 [27:12:48<30:49, 142.25s/it]

(18275, 13)


 98%|████████████████████████████████████▍| 738/750 [27:15:08<28:18, 141.51s/it]

(18300, 13)


 99%|████████████████████████████████████▍| 739/750 [27:17:30<26:00, 141.89s/it]

(18325, 13)


 99%|████████████████████████████████████▌| 740/750 [27:19:51<23:33, 141.35s/it]

(18350, 13)


 99%|████████████████████████████████████▌| 741/750 [27:22:12<21:11, 141.26s/it]

(18375, 13)


 99%|████████████████████████████████████▌| 742/750 [27:24:32<18:48, 141.06s/it]

(18400, 13)


 99%|████████████████████████████████████▋| 743/750 [27:26:53<16:27, 141.06s/it]

(18425, 13)


 99%|████████████████████████████████████▋| 744/750 [27:29:14<14:05, 140.87s/it]

(18450, 13)


 99%|████████████████████████████████████▊| 745/750 [27:31:37<11:47, 141.46s/it]

(18475, 13)


 99%|████████████████████████████████████▊| 746/750 [27:34:04<09:33, 143.38s/it]

(18500, 13)


100%|████████████████████████████████████▊| 747/750 [27:36:25<07:07, 142.53s/it]

(18525, 13)


100%|████████████████████████████████████▉| 748/750 [27:38:41<04:41, 140.65s/it]

(18550, 13)


100%|████████████████████████████████████▉| 749/750 [27:41:01<02:20, 140.54s/it]

(18575, 13)


100%|█████████████████████████████████████| 750/750 [27:43:20<00:00, 133.07s/it]


# Evaluating and cleaning the data

I will import again the data to have a save point for the raw data.

In [24]:
df_0711_zoopla = pd.read_csv("df_0711_zoopla.csv", index_col = 0)

## Checking raw data and duplicates

Within my first scraping round i received a result of 47.917 results before i cleaned duplicates. Over time the amount of objects will increase as i am scraping Zoopla every month. The below shape will not agree as it just reflects the last scraping. 

In [25]:
df_zoopla = df_0711_zoopla
df_zoopla.shape

(18575, 13)

To avoid any duplicates i will directly drop the duplicates within the scraped data. 

In [26]:
df_zoopla = df_zoopla.drop_duplicates()
print(df_zoopla.shape)

(18575, 13)


## Data cleaning 

For the data cleaning I will have a look at each column and identify if and how the information would need to be adjusted to make them usable for EDA and modeling.  

In [27]:
df_zoopla.columns

Index(['name', 'location', 'price', 'bedrooms', 'bathrooms', 'receptions',
       'published', 'seize', 'public_transport', 'describtion', 'links',
       'latitude', 'longitude'],
      dtype='object')

The dataFrame includes the following columns
   - title (here called "name")
   - location
   - price
   - bedrooms
   - bathrooms
   - receptions
   - published
   - size
   - public_transport
   - description
   - links
   - latitude
   - longitude
    

In [28]:
# renaming existing columns if needed 

df_zoopla.columns = ['name', 'location', 'price', 'bedrooms', 'bathrooms', 'receptions',
       'published', 'size', 'public_transport', 'describtion', 'links',
       'latitude', 'longitude']

### name

I will not adjust anything for the name column

### location 

First I will extract the postcode and saved it within an extra column.

In [29]:
# Function to extract the postcode
def extract_postcode (i):
    try:
        match = re.search('([A-Z]+[0-9]+)',i)
        return (i[match.start():match.end()])
    except:
        return np.nan

In [30]:
df_zoopla["postcode"] = df_zoopla["location"].apply(extract_postcode)

**Google API for longitude and latitude**

Within the first scraping i used the google API to collect the latitude and longitude for the apartments. For the runs afterwards i updated the scraper so that the longitude and latitude are directly extracted from Zoopla. As a result this step is not needed anymore for new data. 

In [31]:
# gmaps = googlemaps.Client(key='AIzaSyA8oJA1J2N1PzpxG_MAj8gnY17RSkYaELA')

The below function is using the location column and requesting from google a longitude and latitude location for this position. 

In [32]:
# for x in range(len(df_zoopla["location"])):
#     try:
#         geocode_result = gmaps.geocode(df_zoopla["location"][x])
#         df_zoopla["latitude"][x] = geocode_result[0]['geometry']['location'] ['lat']
#         df_zoopla["longitude"][x] = geocode_result[0]['geometry']['location']['lng']
#     except:
#         df_zoopla["latitude"][x] = np.nan
#         df_zoopla["longitudeg"][x] = np.nan

### price

The price columns would need some adjustment and cleaning I only want to have the price per mount as integer included.

In [33]:
df_zoopla["price"] = df_zoopla["price"].str.replace("£", "")
df_zoopla["price"] = df_zoopla["price"].str.replace(",", "")
df_zoopla["price"] = df_zoopla["price"].str.replace("From", "")
df_zoopla["price"] = df_zoopla["price"].str.replace("Guide price", "")
df_zoopla["price"] = df_zoopla["price"].str.replace("Fixed price", "")
df_zoopla["price"] = df_zoopla["price"].str.replace("Offers over", "")
df_zoopla["price"] = df_zoopla["price"].replace(to_replace ="(pcm).*(pw)", value = '', regex = True)
df_zoopla["price"] = df_zoopla["price"].str.replace("Offers in region of", "")
df_zoopla["price"] = df_zoopla["price"].str.replace("POA", "")

Afterwards we will covert to values into numeric values.

In [37]:
df_zoopla["price"] = df_zoopla["price"].apply(pd.to_numeric) 

### bedrooms, bathrooms and receptions

All three columns would need some adjustments as I only want to have numeric values for all three. 

In [38]:
df_zoopla["bedrooms"] = df_zoopla["bedrooms"].str.replace("bed", "")
df_zoopla["bedrooms"] = df_zoopla["bedrooms"].str.replace("-", "")
df_zoopla["bedrooms"] = df_zoopla["bedrooms"].str.replace("s", "")
df_zoopla["bathrooms"] = df_zoopla["bathrooms"].str.replace("bath", "")
df_zoopla["bathrooms"] = df_zoopla["bathrooms"].str.replace("-", "")
df_zoopla["bathrooms"] = df_zoopla["bathrooms"].str.replace("s", "")
df_zoopla["receptions"] = df_zoopla["receptions"].str.replace("reception", "")
df_zoopla["receptions"] = df_zoopla["receptions"].str.replace("-", "")
df_zoopla["receptions"] = df_zoopla["receptions"].str.replace("s", "")

After cleaning the data we will convert the values into numeric information.

In [39]:
df_zoopla["bedrooms"] = df_zoopla["bedrooms"].apply(pd.to_numeric) 
df_zoopla["bathrooms"] = df_zoopla["bathrooms"].apply(pd.to_numeric) 
df_zoopla["receptions"] = df_zoopla["receptions"].apply(pd.to_numeric) 

### size

Unfortunately not for a lot of apartments the size was included. Nevertheless we will adjust the values to only include numeric values which represent sq. ft.

In [42]:
df_zoopla["size"] = df_zoopla["size"].str.replace("sq.", "")
df_zoopla["size"] = df_zoopla["size"].str.replace(",", "")
df_zoopla["size"] = df_zoopla["size"].str.replace("ft", "")

In [41]:
df_zoopla["size"] = df_zoopla["size"].apply(pd.to_numeric) 

### published

For potential later projects I will convert the published column into the format of a datetime. Like this i can always keep track when the object had been published. 

In [43]:
df_zoopla["published"] = df_zoopla["published"].str.replace("Listed on ", "")

In [44]:
df_zoopla['published'] = pd.to_datetime(df_zoopla['published'])

### location_long and location_lat

I would like to be sure that the objects are only located within London. I would like to conclude the outer ring of London as border for results. I reviewed the latitude and longitude of this on google maps and concluded that the borders should be as follows: 
- longitude should be between 51.264825 and 51.6877167
- latitude should be between -0.498934 and 0.271996 

In [45]:
max_lat = 51.6877167
min_lat = 51.264825

max_long = 0.271996
min_long = -0.498934

In [46]:
df_zoopla = df_zoopla.drop(df_zoopla[(df_zoopla["longitude"]> max_long) | (df_zoopla["longitude"]< min_long)].index)

In [47]:
df_zoopla = df_zoopla.drop(df_zoopla[(df_zoopla["latitude"]> max_lat) | (df_zoopla["latitude"]< min_lat)].index)

### public_transport, describtion and links

For now i will not adjust any of the information in here. For the public transport i will actually use another Dataset and calculate the distance between the objects and the stations to receive this information. 

# Combining different results of the scraper

In a first step i will save the scraped and cleaned data.

In [48]:
df_zoopla.to_csv(f'df_0711_zoopla.csv')

I would like to run the scraper several times. I expect that after a month new objects will be published on the website. Over the next months i will run the scraper each month and increase the amount of data. Every time I will save the single dataframes. Within this section i will load them and combine them to one dataframe. 

In [49]:
df_zoopla_old = pd.read_csv("df_zoopla_final.csv", index_col = 0) # always taking the full final dataset from last time 
df_0711_zoopla = pd.read_csv("df_0711_zoopla.csv", index_col = 0)

In [50]:
print(df_zoopla_old.shape)
print(df_0711_zoopla.shape)

(65052, 14)
(18529, 14)


In [51]:
df_zoopla_final = pd.concat([df_0711_zoopla, df_zoopla_old])
df_zoopla_final.shape

(83581, 14)

At this point it is very important to drop the duplicates. It can be that some of the objects which had been listed a month ago are still listed on the website. For this reason i will drop the duplicates.  

In [52]:
df_zoopla_final = df_zoopla_final.drop_duplicates()
print(df_zoopla_final.shape)

(83581, 14)


Check for the type of data in each column:

In [53]:
df_zoopla_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 83581 entries, 0 to 47916
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              83581 non-null  object 
 1   location          83581 non-null  object 
 2   price             83573 non-null  float64
 3   bedrooms          71161 non-null  float64
 4   bathrooms         77240 non-null  float64
 5   receptions        56284 non-null  float64
 6   published         83581 non-null  object 
 7   size              4467 non-null   float64
 8   public_transport  83479 non-null  object 
 9   describtion       83511 non-null  object 
 10  links             83581 non-null  object 
 11  latitude          83475 non-null  float64
 12  longitude         83475 non-null  float64
 13  postcode          83581 non-null  object 
dtypes: float64(7), object(7)
memory usage: 9.6+ MB


# Plotting the data into a map

In [55]:
map1 = KeplerGl(height=600, width=800) #show the map
# Create a geodataframe
gdf = gpd.GeoDataFrame(df_zoopla, geometry=gpd.points_from_xy(df_zoopla.longitude, df_zoopla.latitude))# Add data to Kepler
map1.add_data(data=gdf, name='kepler map') # adding geo enabled dataframe to map
map1

User Guide: https://docs.kepler.gl/docs/keplergl-jupyter


KeplerGl(data={'kepler map': {'index': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, …

# Saving the data to a final dataframe for the next step EDA 

In [56]:
df_zoopla_final.to_csv(f'df_zoopla_final.csv')